---
title: "Annotations"
author: Ahmed Hasan
output: pdf_document
---

To do:

1. Get proportion of new muts in CDS/UTRs/introns/intergenic across groups + types
2. Get proportion of new muts in the same annotations for salt/expressed gene sets
3. Generate bootstrap confidence intervals
4. Repeat all of the above for MA/adaptation datasets! 

# Import packages and load in data

```{r}
library(tidyverse)
library(fs)
library(here)
library(boot)
library(tictoc)
library(janitor)
library(wesanderson)
```

## Mutation datasets 

Loading in the 'final' versions with gene set info columns:

```{r}
salt_ma = read_tsv(
  here('data/rate/mut_describer/muts_described.final.gene_sets.tsv'),
  col_types = cols()
) %>% 
  mutate(rho = as.double(rho)) %>% 
  bind_rows(
    read_tsv(
      here('data/rate/mut_describer/indels_described.gene_sets.tsv'),
      col_types = cols()
    ) %>% 
      mutate(mut_codon = as.character(mut_codon)) # needed for join
  ) %>% 
  mutate(
    sample_group = str_extract(mutant_sample, '^[A-Z]{2}'),
    sample_type = str_extract(mutant_sample, '[05]$'),
    sample_factor = paste(sample_group, sample_type, sep = '_')
  )

ma_original <- read_tsv(
  here('data/rate/mut_describer/ma.gene_sets.tsv'), 
  col_types = cols(), na = c('.')
) %>% 
  select(-chromosome_1, -position_1)

# splice sites for silent/nonsilent classification of HS muts
splice_sites = data.frame(
  chromosome = c('chromosome_12', 'chromosome_5', 'chromosome_6', 'chromosome_8'),
  position = c(8771346, 2315498, 2122297, 1131421)
)

adaptation = read_tsv(
  here('data/rate/mut_describer/adaptation.gene_sets.tsv'),
  col_types = cols(), na = c('.')
) %>% 
  mutate(site_type = case_when(
    Nonsyn_V_Syn == 'synonymous' | intronic == 1 | intergenic == 1 ~ 'silent',
    TRUE ~ 'nonsilent'
  )) %>% 
  mutate(site_type = ifelse(
    chromosome %in% splice_sites$chromosome & position %in% splice_sites$position, 'nonsilent', site_type
  ))
```

## Callables files

```{r}
# saltMA callables datasets
annotation_callables = read_tsv(
  here('data/rate/annotation_callables.tsv'),
  col_types = cols())

salt_callables = read_tsv(
  here('data/rate/annotation_callables_samples.tsv'),
  col_types = cols()
) %>% 
  mutate(
    sample_group = str_extract(sample, '^[A-Z]{2}'),
    sample_type = str_extract(sample, '[05]$'),
    sample_factor = paste(sample_group, sample_type, sep = '_')) %>% 
  group_by(sample_factor, scaffold, annotation) %>% 
  summarise(
    sample_count = length(unique(sample)),
    callable_total = sum(callable_total),
    genome_count = max(genome_count),
    .groups = 'drop'
  ) 

salt_sample_groups = salt_callables %>% 
  select(sample_factor, sample_count) %>% 
  distinct()


# original MA callables
ma_callables = read_csv(
  here('data/prev/annotation.csv'),
  col_types = cols()
) %>% 
  janitor::clean_names() %>% 
  select(-type, -count) %>% 
  distinct() %>% 
  rename(
    annotation = category,
    callable_total = total_callable)
  
# adaptation callables
adaptation_callables = read_csv(
  here('data/rate/adaptation_callables.csv'),
  col_types = cols()
) %>% 
  # cleanup
  janitor::clean_names() %>% 
  select(-x1) %>% 
  # transform
  pivot_longer(
    cds:intronic,
    names_to = 'annotation',
    values_to = 'callables'
  ) %>% 
  mutate(annotation = ifelse(
    annotation == 'cds', 'CDS', annotation)) %>% 
  group_by(annotation) %>% 
  summarise(callable_total = sum(callables))
```

# Mutation proportions

## Salt MA

```{r}
props_all = list()

props_all$salt_ma = salt_ma %>% 
  group_by(sample_factor) %>% 
  select(
    sample_factor, sample_group, sample_type,
    CDS, utr5, utr3, intronic, intergenic
  ) %>% 
  
  # get n_muts per annotation per sample factor
  summarise(
    across(
      where(is.numeric), sum)
    ) %>% 
  
  pivot_longer(
    cols = where(is.numeric),
    names_to = 'annotation',
    values_to = 'values'
  ) %>% 
  
  # add callable counts
  left_join(
    salt_callables %>% 
      group_by(sample_factor, annotation) %>% 
      summarise(
        callable_total = sum(callable_total),
        sample_count = max(sample_count),
        .groups = 'drop'
      ),
    by = c('sample_factor', 'annotation')
  ) %>% 
  
  # fix sample counts
  select(-sample_count) %>% 
  left_join(salt_sample_groups, by = 'sample_factor') %>% 
  group_by(sample_factor) %>% 
  mutate(
    numerator = (values / callable_total),
    denominator = sum(numerator),
    prop = numerator / denominator
  )

props_all$salt_ma
```

## MA and Adaptation

```{r}
get_props = function(d, d_callables, salt_ma_bool = FALSE) {
  out = d %>% 
    { if (salt_ma_bool == TRUE) group_by(., sample_factor) else . } %>% 
    select(
      sample_factor, CDS, utr5, utr3, intronic, intergenic
    ) %>% 
    
    # get n_muts per annotation
    summarise(
      across(
        where(is.numeric), sum),
      sample_factor = unique(sample_factor)
      ) %>% 
    
    pivot_longer(
      cols = where(is.numeric),
      names_to = 'annotation',
      values_to = 'values'
    )
  
  # add callable counts
  if (salt_ma_bool == TRUE) {
    out = out %>% 
      left_join(
        d_callables %>% 
          group_by(sample_factor, annotation) %>% 
          summarise(
            callable_total = sum(callable_total),
            .groups = 'drop'
          ),
        by = c('sample_factor', 'annotation')
      ) 
  } else if (salt_ma_bool == FALSE) {
    out = out %>% 
      left_join(
        d_callables %>% 
          group_by(annotation) %>% 
          summarise(
            callable_total = sum(callable_total),
            .groups = 'drop'
          ),
        by = c('annotation')
      )
  }
      
  out = out %>% 
    # fix sample counts
    mutate(
      numerator = (values / callable_total),
      denominator = sum(numerator, na.rm = TRUE),
      prop = numerator / denominator
    )
  return(out)
}

props_all$adaptation = get_props(
  adaptation %>% 
    mutate(sample_factor = 'adaptation'), adaptation_callables)
props_all$ma_original = get_props(
  ma_original %>% 
    mutate(sample_factor = 'ma_original'), ma_callables)
```

## Salt MA summarised

Splitting by 0 and 5:


```{r}
salt_callables_summarised = salt_callables %>% 
  mutate(sample_factor = str_extract(sample_factor, '[05]$')) %>% 
  mutate(sample_factor = case_when(
    sample_factor == '0' ~ 'salt 0',
    sample_factor == '5' ~ 'salt 5')
  ) %>% 
  group_by(sample_factor, scaffold, annotation) %>% 
  summarise(
    callable_total = sum(callable_total), 
    genome_count = unique(genome_count), 
    .groups = 'drop')
  
props_all$salt_ma_0_all = get_props(
  salt_ma %>% 
    filter(sample_type == 0) %>% 
    mutate(sample_factor = 'salt 0'),
  salt_callables_summarised %>% 
    filter(sample_factor == 'salt 0')
)

props_all$salt_ma_5_all = get_props(
  salt_ma %>% 
    filter(sample_type == 5) %>% 
    mutate(sample_factor = 'salt 5'),
  salt_callables_summarised %>% 
    filter(sample_factor == 'salt 5')
)

```



# Bootstrap CIs

```{r}
get_prop_value = function(d, indices, ant, d_callables, salt_ma_bool = FALSE) {
  d_resampled = d[indices, ]
  prop_value = get_props(d_resampled, d_callables, salt_ma_bool = salt_ma_bool) %>% 
    filter(annotation == ant) %>% 
    select(prop) %>% 
    as.numeric()
  return(prop_value)
}

prop_bootstrap = function(d, ant, d_callables, rep_count, salt_ma_bool = FALSE) {
  cis_all = d %>% 
    split(.$sample_factor) %>% 
    map_dfr(., function(d_sub) {
      tic(msg = paste('bootstrap for', unique(d_sub$sample_factor), ant))
      message(paste('starting', unique(d_sub$sample_factor), ant))
      
      # check if prop is 0
      prop_val = get_prop_value(
        d_sub, c(1:nrow(d_sub)), ant, 
        d_callables, salt_ma_bool = salt_ma_bool)
      
      # no CIs if proportion is 0
      if (prop_val %in% c(0, 1)) {
        message('prop val is either 0 or 1 - CIs cannot be calculated')
        cis = c(0, 0)
      } else {
        message(paste('bootstrapping - prop val is', prop_val))
        boot_obj = boot(data = d_sub, statistic = get_prop_value,
          R = rep_count, ant = ant, d_callables = d_callables, salt_ma_bool = salt_ma_bool)
        cis = boot.ci(boot_obj, type = 'perc')$perc[c(4, 5)]
      }
      
      # prep data frame with CIs to return
      out = data.frame(
        ant = ant, lower_ci = cis[1], upper_ci = cis[2])
      toc(log = TRUE)
      return(out)
    },
    .id = 'sample_factor'
    )
  return(cis_all)
}

prop_bootstrap(adaptation %>% mutate(sample_factor = 'adaptation'), 'CDS', 
               adaptation_callables, 500, salt_ma_bool = FALSE)
```

Running over all possible sample combinations and annotations (CC and DL)

```{r}
annotations = list('CDS', 'utr5', 'utr3', 'intronic', 'intergenic')

salt_ma_cis = annotations %>% 
  map_dfr(
    ~ prop_bootstrap(
        salt_ma %>% filter(sample_group != 'SL'), ., salt_callables, 1000, salt_ma_bool = TRUE)
  ) 

salt_ma_cis

salt_ma_prop_all = full_join(
  props_all$salt_ma,
  salt_ma_cis %>% 
    rename(annotation = ant), 
  by = c('sample_factor', 'annotation')
) %>% 
  select(sample_factor, annotation, prop, contains('ci'), everything())

salt_ma_prop_all

# doing the SLs 
salt_ma_SL_cis = annotations %>% 
  map_dfr(
    ~ prop_bootstrap(
        salt_ma %>% filter(sample_group == 'SL'), ., salt_callables, 1000, salt_ma_bool = TRUE)
  )

salt_ma_prop_all = left_join(
  salt_ma_prop_all, 
  salt_ma_SL_cis %>% rename(annotation = ant), 
  by = c('sample_factor', 'annotation')) %>% 
  mutate(
    lower_ci.x = ifelse(is.na(lower_ci.x), lower_ci.y, lower_ci.x), 
    upper_ci.x = ifelse(is.na(upper_ci.x), upper_ci.y, upper_ci.x)) %>% 
  select(-lower_ci.y, -upper_ci.y) %>% 
  rename(lower_ci = lower_ci.x, upper_ci = upper_ci.x)
  
# write_tsv(salt_ma_prop_all, here('data/spatial/salt_ma_annotations_cis.tsv'))

```

Repeating for 0 and 5 summarised:

```{r}
# salt 0
salt_0_cis = map_dfr(
  annotations,
  ~ prop_bootstrap(
    salt_ma %>% 
      filter(sample_type == 0) %>% 
      mutate(sample_factor = 'salt 0'),
    .,
    salt_callables_summarised %>% 
      filter(sample_factor == 'salt 0'),
    1000, salt_ma_bool = FALSE
  )
)

salt_5_cis = map_dfr(
  annotations,
  ~ prop_bootstrap(
    salt_ma %>% 
      filter(sample_type == 5) %>% 
      mutate(sample_factor = 'salt 5'),
    .,
    salt_callables_summarised %>% 
      filter(sample_factor == 'salt 5'),
    1000, salt_ma_bool = FALSE
  )
)

salt_0_all = full_join(
  props_all$salt_ma_0_all,
  salt_0_cis %>% 
    rename(annotation = ant),
  by = c('sample_factor', 'annotation')
) %>% 
  select(sample_factor, annotation, prop, contains('ci'), everything())

salt_5_all = full_join(
  props_all$salt_ma_5_all,
  salt_5_cis %>% 
    rename(annotation = ant),
  by = c('sample_factor', 'annotation')
) %>% 
  select(sample_factor, annotation, prop, contains('ci'), everything())

# write_tsv(salt_0_all, here('data/spatial/salt_0_all_annotations_cis.tsv'))
# write_tsv(salt_5_all, here('data/spatial/salt_5_all_annotations_cis.tsv'))


```


Repeating for MA and adaptation:

```{r}
# adaptation
adaptation_cis = map_dfr(
  annotations,
  ~ prop_bootstrap(
      adaptation %>% 
        mutate(sample_factor = 'adaptation'), 
      ., adaptation_callables, 1000, salt_ma_bool = FALSE
  )
)

adaptation_prop_all = full_join(
  props_all$adaptation,
  adaptation_cis %>% 
    rename(annotation = ant),
  by = c('sample_factor', 'annotation')
) %>% 
  select(sample_factor, annotation, prop, contains('ci'), everything())

adaptation_prop_all

# write_tsv(adaptation_prop_all, here('data/spatial/adaptation_annotations_cis.tsv'))

# ma
ma_cis = map_dfr(
  annotations,
  ~ prop_bootstrap(
    ma_original %>% 
      mutate(sample_factor = 'ma_original'),
    ., ma_callables, 1000, salt_ma_bool = FALSE
  )
)

ma_prop_all = full_join(
  props_all$ma_original,
  ma_cis %>% 
    rename(annotation = ant),
  by = c('sample_factor', 'annotation')
) %>% 
  select(sample_factor, annotation, prop, contains('ci'), everything())

ma_prop_all

# write_tsv(ma_prop_all, here('data/spatial/ma_annotations_cis.tsv'))
```


## Plot

```{r}
salt_ma_prop_all = read_tsv(
  here('data/spatial/salt_ma_annotations_cis.tsv'),
  col_types = cols()
) %>% 
  mutate(
    sample_type = str_extract(sample_factor, '^[A-Z]{2}'),
    sample_group = str_extract(sample_factor, '[05]$')
  )

salt_ma_prop_all %>% 
  mutate(sample_type = as.character(sample_type)) %>% 
  filter(sample_group != 'SL') %>% 
  
  ggplot(aes(x = annotation, y = prop, fill = sample_type)) +
  geom_bar(stat = 'identity', position = position_dodge(width = 0.9)) +
  facet_wrap(~ sample_group) +
  geom_errorbar(
    aes(ymin = lower_ci, ymax = upper_ci, group = sample_type),
    position = position_dodge(width = 0.9),
    size = 0.5, width = 0.25) +
  theme_classic() +
  theme(
    axis.text = element_text(size = 12, color = 'black'),
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.title = element_text(size = 12, color = 'black'),
    legend.title = element_blank(),
    # legend.position = c(0.2, 0.88),
    legend.position = 'top',
    legend.background = element_blank(),
    legend.text = element_text(size = 10, color = 'black')
  ) +
  coord_cartesian(
    y = c(0, 1)
  ) +
  scale_x_discrete(
    limits = c('CDS', 'utr5', 'utr3', 'intronic', 'intergenic')
  )
  
```

With 0 and 5 summarised:

```{r}
palette = wes_palette(7, name = 'Darjeeling1', type = 'continuous')

salt_ma_prop_summarised = bind_rows(
  read_tsv(here('data/spatial/salt_0_all_annotations_cis.tsv'), col_types = cols()),
  read_tsv(here('data/spatial/salt_5_all_annotations_cis.tsv'), col_types = cols())
)

salt_ma_prop_summarised %>% 
  ggplot(aes(x = annotation, y = prop, fill = sample_factor)) +
  geom_bar(
    stat = 'identity', 
    color = 'black',
    position = position_dodge(width = 0.9)) +
  geom_errorbar(
    aes(ymin = lower_ci, ymax = upper_ci, group = sample_factor),
    position = position_dodge(width = 0.9),
    size = 0.5, width = 0.25) +
  theme_classic() +
  theme(
    axis.text = element_text(size = 12, color = 'black'),
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.title = element_text(size = 12, color = 'black'),
    legend.title = element_blank(),
    # legend.position = c(0.2, 0.88),
    legend.position = 'top',
    legend.background = element_blank(),
    legend.text = element_text(size = 10, color = 'black')
  ) +
  coord_cartesian(
    y = c(0, 1)
  ) +
  scale_x_discrete(
    limits = c('CDS', 'utr5', 'utr3', 'intronic', 'intergenic')
  ) +
  scale_fill_manual(
    values = palette[c(6, 7)]
  )
```



# Mutation proportions - for gene sets

## Salt MA

Load in data:

```{r}
# mutation properties
salt_ma_gene_sets = read_tsv(
  here('data/rate/mut_describer/muts_described.final.gene_sets.tsv'),
  col_types = cols()) %>% 
  mutate(
    sample_group = str_extract(mutant_sample, '^[A-Z]{2}'),
    sample_type = str_extract(mutant_sample, '[05]$'),
    sample_factor = paste(sample_group, sample_type, sep = '_')
  ) %>% 
  bind_rows(
    read_tsv(
      here('data/rate/mut_describer/indels_described.gene_sets.tsv'),
      col_types = cols()
    ) %>% 
      mutate(mut_codon = as.character(mut_codon),
             rho = as.character(rho))
  )
  
expressed_muts = list()
perrineau_muts = list()

expressed_muts$salt_ma = salt_ma_gene_sets %>% 
  filter(expressed == 1)
expressed_muts$adaptation = adaptation %>% 
  filter(expressed == 1) %>% 
  mutate(sample_factor = 'adaptation')
expressed_muts$ma_orig = ma_original %>% 
  filter(expressed == 1) %>% 
  mutate(sample_factor = 'ma_original')

perrineau_muts$salt_ma = salt_ma_gene_sets %>% 
  filter(perrineau == 1)
perrineau_muts$adaptation = adaptation %>% 
  filter(perrineau == 1) %>% 
  mutate(sample_factor = 'adaptation')
perrineau_muts$ma_orig = ma_original %>% 
  filter(perrineau == 1) %>% 
  mutate(sample_factor = 'ma_original')

```


Load in callables files:

```{r}
callables = list()

callables$salt_ma_expressed = read_tsv(
  here('data/rate/gene_lists/expressed_callables_summarised.tsv'),
  col_types = cols()
) %>% 
  rename(callable_total = callables_total)
callables$salt_ma_perrineau = read_tsv(
  here('data/rate/gene_lists/perrineau_callables_summarised.tsv'),
  col_types = cols()
) %>% 
  rename(callable_total = callables_total)

callables$ma_orig_expressed = read_tsv(
  here('data/rate/gene_lists/expressed_ma_callables_summarised.tsv'),
  col_types = cols()
) %>% 
  mutate(sample_factor = 'ma_original') %>% 
  rename(callable_total = callables_total)
callables$ma_orig_perrineau = read_tsv(
  here('data/rate/gene_lists/perrineau_ma_callables_summarised.tsv'),
  col_types = cols()
) %>% 
  mutate(sample_factor = 'ma_original') %>% 
  rename(callable_total = callables_total)

callables$adaptation_expressed = read_csv(
  here('data/prev/summary_callables_expressed.csv'),
  col_types = cols()
) %>% 
  pivot_longer(
    CDS:intronic, names_to = 'annotation', values_to = 'callables_total'
  ) %>% 
  group_by(annotation) %>% 
  summarise(callable_total = sum(callables_total)) %>% 
  mutate(sample_factor = 'adaptation')
callables$adaptation_perrineau = read_csv(
  here('data/prev/summary_salt_gene_callables.csv'),
  col_types = cols()
) %>% 
  pivot_longer(
    CDS:intronic, names_to = 'annotation', values_to = 'callables_total'
  ) %>% 
  group_by(annotation) %>% 
  summarise(callable_total = sum(callables_total)) %>% 
  mutate(sample_factor = 'adaptation')

# all the callables need intergenic = 0 for compatibility with earlier fxns
callables[c(1:4)] = map(callables[c(1:4)], 
  ~ bind_rows(.,
    data.frame(
      sample_factor = unique(.$sample_factor),
      annotation = 'intergenic',
      callable_total = 0)
  )
)
```

Get proportions using fxns above:

```{r}
get_props_salt_ma = function(d, d_callables) {
  d %>% 
    group_by(sample_factor) %>% 
    select(
      sample_factor, sample_group, sample_type,
      CDS, utr5, utr3, intronic, intergenic
    ) %>% 
    
    # get n_muts per annotation per sample factor
    summarise(
      across(
        where(is.numeric), sum)
      ) %>% 
    
    pivot_longer(
      cols = where(is.numeric),
      names_to = 'annotation',
      values_to = 'values'
    ) %>% 
    
    # add callable counts
    left_join(
      d_callables %>% 
        group_by(sample_factor, annotation) %>% 
        summarise(
          callable_total = sum(callable_total),
          .groups = 'drop'
        ),
      by = c('sample_factor', 'annotation')
    ) %>% 
    group_by(sample_factor) %>% 
    mutate(
      numerator = (values / callable_total),
      denominator = sum(numerator, na.rm = TRUE),
      prop = numerator / denominator
    )
}
  
props_expressed = list()
props_perrineau = list()

# salt MA
props_expressed$salt_ma = get_props_salt_ma(
  expressed_muts$salt_ma, 
  callables$salt_ma_expressed)

props_perrineau$salt_ma = get_props_salt_ma(
  perrineau_muts$salt_ma,
  callables$salt_ma_perrineau
)

# adaptation
props_expressed$adaptation = get_props(
  expressed_muts$adaptation,
  callables$adaptation_expressed,
  salt_ma_bool = FALSE
)

props_perrineau$adaptation = get_props(
  perrineau_muts$adaptation,
  callables$adaptation_perrineau,
  salt_ma_bool = FALSE
)

# MA original
props_expressed$ma_orig = get_props(
  expressed_muts$ma_orig,
  callables$ma_orig_expressed,
  salt_ma_bool = FALSE
)

props_perrineau$ma_orig = get_props(
  perrineau_muts$ma_orig,
  callables$ma_orig_perrineau,
  salt_ma_bool = FALSE
)

# salt MA summarised
props_expressed$salt_ma_0 = get_props_salt_ma(
  expressed_muts$salt_ma %>% 
    filter(sample_type == 0) %>% 
    mutate(sample_factor = 'salt 0'),
  callables$salt_ma_expressed %>% 
    mutate(sample_type = str_extract(sample_factor, '[05]$')) %>% 
    filter(sample_type == 0) %>% 
    group_by(annotation) %>% 
    summarise(callable_total = sum(callable_total)) %>% 
    mutate(sample_factor = 'salt 0')
)

props_expressed$salt_ma_5 = get_props_salt_ma(
  expressed_muts$salt_ma %>% 
    filter(sample_type == 5) %>% 
    mutate(sample_factor = 'salt 5'),
  callables$salt_ma_expressed %>% 
    mutate(sample_type = str_extract(sample_factor, '[05]$')) %>% 
    filter(sample_type == 5) %>% 
    group_by(annotation) %>% 
    summarise(callable_total = sum(callable_total)) %>% 
    mutate(sample_factor = 'salt 5')
)

props_perrineau$salt_ma_0 = get_props_salt_ma(
  perrineau_muts$salt_ma %>% 
    filter(sample_type == 0) %>% 
    mutate(sample_factor = 'salt 0'),
  callables$salt_ma_perrineau %>% 
    mutate(sample_type = str_extract(sample_factor, '[05]$')) %>% 
    filter(sample_type == 0) %>% 
    group_by(annotation) %>% 
    summarise(callable_total = sum(callable_total)) %>% 
    mutate(sample_factor = 'salt 0')
)

props_perrineau$salt_ma_5 = get_props_salt_ma(
  perrineau_muts$salt_ma %>% 
    filter(sample_type == 5) %>% 
    mutate(sample_factor = 'salt 5'),
  callables$salt_ma_perrineau %>% 
    mutate(sample_type = str_extract(sample_factor, '[05]$')) %>% 
    filter(sample_type == 5) %>% 
    group_by(annotation) %>% 
    summarise(callable_total = sum(callable_total)) %>% 
    mutate(sample_factor = 'salt 5')
)

# remove intergenic lines
props_expressed = map(props_expressed, ~ filter(., annotation != 'intergenic'))
props_perrineau = map(props_perrineau, ~ filter(., annotation != 'intergenic'))

```


## Bootstrap CIs

```{r}
annotations_gene_sets = list('CDS', 'utr5', 'utr3', 'intronic')

expressed_cis = list()
perrineau_cis = list()

# expressed genes
expressed_cis$salt_ma = annotations_gene_sets %>% 
  map_dfr(
    ~ prop_bootstrap(
      expressed_muts$salt_ma, ., callables$salt_ma_expressed, 1000, salt_ma_bool = TRUE
    )
  ) 

expressed_cis$adaptation = annotations_gene_sets %>% 
  map_dfr(
    ~ prop_bootstrap(
      expressed_muts$adaptation, ., callables$adaptation_expressed, 1000, salt_ma_bool = FALSE
    )
  )

expressed_cis$ma_orig = annotations_gene_sets %>% 
  map_dfr(
    ~ prop_bootstrap(
      expressed_muts$ma_orig, ., callables$ma_orig_expressed, 1000, salt_ma_bool = FALSE
    )
  )

# perrineau
perrineau_cis$salt_ma = annotations_gene_sets %>% 
  map_dfr(
    ~ prop_bootstrap(
      perrineau_muts$salt_ma, ., callables$salt_ma_perrineau, 1000, salt_ma_bool = TRUE
    )
  )

perrineau_cis$adaptation = annotations_gene_sets %>% 
  map_dfr(
    ~ prop_bootstrap(
      perrineau_muts$adaptation, ., callables$adaptation_perrineau, 1000, salt_ma_bool = FALSE
    )
  )

perrineau_cis$ma_orig = annotations_gene_sets %>% 
  map_dfr(
    ~ prop_bootstrap(
      perrineau_muts$ma_orig, ., callables$ma_orig_perrineau, 1000, salt_ma_bool = FALSE
    )
  )

```

Repeat for salt 0 and 5 combined:

```{r}
# expressed
expressed_cis$salt_ma_0 = annotations_gene_sets %>% 
  map_dfr(
    ~ prop_bootstrap(
      expressed_muts$salt_ma %>% 
        filter(sample_type == 0) %>% 
        mutate(sample_factor = 'salt 0'),
      .,
      callables$salt_ma_expressed %>% 
        mutate(sample_type = str_extract(sample_factor, '[05]$')) %>% 
        filter(sample_type == 0) %>% 
        group_by(annotation) %>% 
        summarise(callable_total = sum(callable_total)) %>% 
        mutate(sample_factor = 'salt 0'),
      1000, salt_ma_bool = FALSE
    )
  )

expressed_cis$salt_ma_5 = annotations_gene_sets %>% 
  map_dfr(
    ~ prop_bootstrap(
      expressed_muts$salt_ma %>% 
        filter(sample_type == 5) %>% 
        mutate(sample_factor = 'salt 5'),
      .,
      callables$salt_ma_expressed %>% 
        mutate(sample_type = str_extract(sample_factor, '[05]$')) %>% 
        filter(sample_type == 5) %>% 
        group_by(annotation) %>% 
        summarise(callable_total = sum(callable_total)) %>% 
        mutate(sample_factor = 'salt 5'),
      1000, salt_ma_bool = FALSE
    )
  )

# perrineau
perrineau_cis$salt_ma_0 = annotations_gene_sets %>% 
  map_dfr(
    ~ prop_bootstrap(
      perrineau_muts$salt_ma %>% 
        filter(sample_type == 0) %>% 
        mutate(sample_factor = 'salt 0'),
      .,
      callables$salt_ma_expressed %>% 
        mutate(sample_type = str_extract(sample_factor, '[05]$')) %>% 
        filter(sample_type == 0) %>% 
        group_by(annotation) %>% 
        summarise(callable_total = sum(callable_total)) %>% 
        mutate(sample_factor = 'salt 0'),
      1000, salt_ma_bool = FALSE
    )
  )

perrineau_cis$salt_ma_5 = annotations_gene_sets %>% 
  map_dfr(
    ~ prop_bootstrap(
      perrineau_muts$salt_ma %>% 
        filter(sample_type == 5) %>% 
        mutate(sample_factor = 'salt 5'),
      .,
      callables$salt_ma_perrineau %>% 
        mutate(sample_type = str_extract(sample_factor, '[05]$')) %>% 
        filter(sample_type == 5) %>% 
        group_by(annotation) %>% 
        summarise(callable_total = sum(callable_total)) %>% 
        mutate(sample_factor = 'salt 5'),
      1000, salt_ma_bool = FALSE
    )
  )

```

Join with originals:

```{r}
expressed_final = list()
perrineau_final = list()

data_groups = c('salt_ma', 'adaptation', 'ma_orig', 'salt_ma_0', 'salt_ma_5')

expressed_final = map(
  data_groups, 
  ~ left_join(
    props_expressed[[.]],
    expressed_cis[[.]] %>% 
      rename(annotation = ant),
    by = c('sample_factor', 'annotation')
  )
) %>% 
  setNames(., data_groups)

perrineau_final = map(
  data_groups,
  ~ left_join(
    props_perrineau[[.]],
    perrineau_cis[[.]] %>% 
      rename(annotation = ant),
    by = c('sample_factor', 'annotation')
  )
) %>% 
  setNames(., data_groups)

# write to file
# write_tsv(expressed_final$salt_ma, here('data/spatial/salt_ma_expressed_cis.tsv'))
# write_tsv(expressed_final$adaptation, here('data/spatial/adaptation_expressed_cis.tsv'))
# write_tsv(expressed_final$ma_orig, here('data/spatial/ma_orig_expressed_cis.tsv'))
# bind_rows(
#   expressed_final$salt_ma_0,
#   expressed_final$salt_ma_5
# ) %>% 
#   write_tsv(., here('data/spatial/salt_ma_summarised_expressed_cis.tsv'))

# write_tsv(perrineau_final$salt_ma, here('data/spatial/salt_ma_perrineau_cis.tsv'))
# write_tsv(perrineau_final$adaptation, here('data/spatial/adaptation_perrineau_cis.tsv'))
# write_tsv(perrineau_final$ma_orig, here('data/spatial/ma_orig_perrineau_cis.tsv'))
# bind_rows(
#   perrineau_final$salt_ma_0,
#   perrineau_final$salt_ma_5
# ) %>%
#   write_tsv(., here('data/spatial/salt_ma_summarised_perrineau_cis.tsv'))
```

## Overall gene set stats

Of the X genic mutations in [data set], Y were in expressed genes - for all salt MA factors (table?)

```{r}
report_props = function(d) {
  total_count = d %>% 
    filter(genic == 1) %>% 
    nrow()
  expressed_count = d %>% 
    filter(expressed == 1) %>% 
    nrow()
  perrineau_count = d %>% 
    filter(perrineau == 1) %>% 
    nrow()
  
  return(c(
    'total' = total_count,
    'expressed' = expressed_count,
    'perrineau' = perrineau_count)
    )
}

gene_set_counts = list()
gene_set_counts$ma_orig = report_props(ma_original)
gene_set_counts$adaptation = report_props(adaptation)
gene_set_counts$adaptation_silent = report_props(adaptation %>% filter(site_type == 'silent'))
gene_set_counts$adaptation_nonsilent = report_props(adaptation %>% filter(site_type == 'nonsilent'))

gene_set_counts = append(
  gene_set_counts,
  map(
    salt_ma_gene_sets %>% split(.$sample_factor),
    ~ report_props(.),
    .id = 'sample_factor'
  )
)

gene_set_counts$salt_0 = gene_set_counts$CC_0 + gene_set_counts$DL_0 + gene_set_counts$SL_0
gene_set_counts$salt_5 = gene_set_counts$CC_5 + gene_set_counts$DL_5 + gene_set_counts$SL_5
```

expressed n_expressed_d1 n_expressed_d2

total n_total_d1 n_total_d2

matrix goes cols first

Chi-square tests to see if proportions of muts in [expressed/salt/etc] are similar to MA proportions

```{r}
do_chisq = function(vec1, vec2) {
  chisq.test(
    matrix(
      c(vec1, vec2),
      nrow = 2, ncol = 2)
  )
}

# expressed genes - MA vs HS
message('expressed genes, MA vs HS')
do_chisq(
  gene_set_counts$adaptation[c('expressed', 'total')], 
  gene_set_counts$ma_orig[c('expressed', 'total')])

# HS - silent vs nonsilent in expressed genes
message('HS expressed, silent vs nonsilent')
do_chisq(
  gene_set_counts$adaptation_silent[c('expressed', 'total')],
  gene_set_counts$adaptation_nonsilent[c('expressed', 'total')]
)

# Perrineau genes - MA vs HS
message('salt genes, MA vs HS')
do_chisq(
  gene_set_counts$adaptation[c('perrineau', 'total')],
  gene_set_counts$ma_orig[c('perrineau', 'total')]
)

# HS - silent vs nonsilent in salt genes
message('HS salt, silent vs nonsilent')
do_chisq(
  gene_set_counts$adaptation_silent[c('perrineau', 'total')],
  gene_set_counts$adaptation_nonsilent[c('perrineau', 'total')]
)

```

Between the Salt MA lines:

```{r}
for (sample_type in list('CC', 'DL', 'SL')) {
  message(paste(sample_type, '0 vs 5'))
  sample_0 = gene_set_counts[[paste(sample_type, '0', sep = '_')]]
  sample_5 = gene_set_counts[[paste(sample_type, '5', sep = '_')]]
  message(paste(sample_type, 'expressed genes'))
  print(do_chisq(
    sample_0[c('expressed', 'total')], sample_5[c('expressed', 'total')]
  ))
  message(paste(sample_type, 'salt genes'))
  print(do_chisq(
    sample_0[c('perrineau', 'total')], sample_5[c('perrineau', 'total')]
  ))
}
```

Between 0, 5, and the previous datasets:

```{r}
# expressed genes - 0 vs 5
names_x = c('salt_0', 'salt_5')
names_y = c('ma_orig', 'adaptation_silent', 'adaptation_nonsilent')

for (name_1 in names_x) {
  d1 = gene_set_counts[[name_1]]
  for (name_2 in names_y) {
    d2 = gene_set_counts[[name_2]]
    message(paste('expressed', name_1, name_2))
    print(do_chisq(
      d1[c('expressed', 'total')],
      d2[c('expressed', 'total')]
    ))
    message(paste('perrineau', name_1, name_2))
    print(do_chisq(
      d1[c('perrineau', 'total')],
      d2[c('perrineau', 'total')]
    ))
  }
}
```


```{r}
message('expressed genes, salt 0 vs 5')
do_chisq(
  gene_set_counts$salt_0[c('expressed', 'total')],
  gene_set_counts$salt_5[c('expressed', 'total')]
)

# Perrineau genes - MA vs HS
message('Perrineau genes, MA vs HS')
do_chisq(
  gene_set_counts$salt_0[c('perrineau', 'total')],
  gene_set_counts$salt_5[c('perrineau', 'total')]
)

# expressed genes - 0 vs MA
message('expressed genes, salt 0 vs MA')
do_chisq(
  gene_set_counts$salt_0[c('expressed', 'total')],
  gene_set_counts$ma_orig[c('expressed', 'total')]
)

# Perrineau genes - 0 vs MA
message('Perrineau genes, salt 0 vs MA')
do_chisq(
  gene_set_counts$salt_0[c('perrineau', 'total')],
  gene_set_counts$ma_orig[c('perrineau', 'total')]
)
```



# Final Plots

Loading in data:

```{r}
load_final = function(dataset) {
  d = map_dfr(
    dir_ls(here('data/spatial'), regexp = paste0('spatial/', dataset, '\\w+')),
    read_tsv, col_types = cols(),
    .id = 'gene_set'
  ) %>% 
  rowwise() %>% 
  mutate(
    gene_set = str_match(gene_set, '(\\w+_\\w+)_cis\\.tsv')[2]
  ) %>% 
  mutate(gene_set = case_when(
    str_detect(gene_set, 'annotation') ~ 'genome wide',
    str_detect(gene_set, 'expressed') ~ 'expressed genes',
    str_detect(gene_set, 'perrineau') ~ 'salt genes'
    )
  )
}

ma_orig_final = load_final('ma')
adaptation_final = load_final('adaptation')
salt_ma_final = load_final('salt_ma') %>% 
  filter(!is.na(prop)) # the odd indel grouping

```

Recreating Fig 2:

```{r}
ma_orig_final %>% 
  bind_rows(adaptation_final) %>% 
  mutate(
    gene_set = factor(
      gene_set, levels = c('genome wide', 'expressed genes', 'salt genes')),
    sample_factor = factor(
      sample_factor, levels = c('ma_original', 'adaptation')),
    annotation = factor(
      annotation, levels = c('CDS', 'utr5', 'utr3', 'intronic', 'intergenic')),
    sample_factor = case_when(
      sample_factor == 'ma_original' ~ 'MA',
      sample_factor == 'adaptation' ~ 'salt-selection'
    )
  ) %>% 
  ggplot(
    aes(x = annotation, y = prop, fill = sample_factor)) +
  geom_bar(
    stat = 'identity', 
    position = position_dodge(width = 0.9),
    colour = 'black',
    size = 0.4) +
  geom_errorbar(
    aes(ymin = lower_ci, ymax = upper_ci),
    position = position_dodge(width = 0.9),
    size = 0.4, width = 0.2
  ) +
  facet_grid(
    ~ gene_set, scales = 'free') +
  coord_cartesian(
    y = c(0, 0.75)
  ) +
  labs(
    x = '',
    y = 'Proportion of all mutations'
  ) +
  scale_fill_manual(
    values = c('MA' = '#87C5FC', 'salt-selection' = '#FF5B5F')
  ) +
  # macro elements
  guides(fill = guide_legend(
    override.aes = list(color = 'white'))
  ) +
  theme(
    panel.border = element_rect(colour = 'black', size = 0.5, fill = NA),
    panel.background = element_blank(),
    strip.background = element_blank(),
    legend.position = c(0.175, 0.85),
    legend.text = element_text(size = 8, colour = 'black'),
    legend.title = element_blank(),
    legend.key = element_rect(color = 'white'),
    legend.key.size = unit(0.5, 'cm')) +
  # text elements
  theme(
    strip.text = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.title = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.text = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )
```

Salt MA data:

```{r}
salt_ma_final %>% 
  mutate(
    gene_set = factor(
      gene_set, levels = c('genome wide', 'expressed genes', 'salt genes')),
    annotation = factor(
      annotation, levels = c('CDS', 'utr5', 'utr3', 'intronic', 'intergenic')),
    sample_factor = str_replace(sample_factor, '_', ' '),
    sample_group = str_extract(sample_factor, '^[A-Z]{2}'),
    sample_type = str_extract(sample_factor, '[05]$')
  ) %>% 
  ggplot(
    aes(x = annotation, y = prop, fill = sample_type)) +
  geom_bar(
    stat = 'identity', 
    position = position_dodge(width = 0.9),
    colour = 'black',
    size = 0.4) +
  geom_errorbar(
    aes(ymin = lower_ci, ymax = upper_ci),
    position = position_dodge(width = 0.9),
    size = 0.4, width = 0.2
  ) +
  facet_grid(
    rows = vars(sample_group), cols = vars(gene_set), scales = 'free') +
  coord_cartesian(
    y = c(0, 1)
  ) +
  labs(
    x = '',
    y = 'Proportion of all mutations'
  ) +
  scale_fill_manual(
    values = c('0' = '#87C5FC', '5' = '#FF5B5F')
  ) +
  # macro elements
  guides(fill = guide_legend(
    override.aes = list(color = 'white'))
  ) +
  theme(
    panel.border = element_rect(colour = 'black', size = 0.5, fill = NA),
    panel.background = element_blank(),
    strip.background = element_blank(),
    legend.text = element_text(size = 8, colour = 'black'),
    legend.title = element_blank(),
    legend.key = element_rect(color = 'white'),
    legend.key.size = unit(0.5, 'cm')) +
  # text elements
  theme(
    strip.text = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.title = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.text = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )
```

Fig 2 style 0 and 5 only plot:

- need to redo Perrineau/expressed across all 0 and all 5

```{r}
palette = wes_palette(7, name = 'Darjeeling1', type = 'continuous')

salt_ma_final_summarised = bind_rows(
  read_tsv(here('data/spatial/salt_0_all_annotations_cis.tsv'), col_types = cols()),
  read_tsv(here('data/spatial/salt_5_all_annotations_cis.tsv'), col_types = cols())
) %>% 
  mutate(gene_set = 'genome wide') %>% 
  bind_rows(
    read_tsv(
      here('data/spatial/salt_ma_summarised_expressed_cis.tsv'),
      col_types = cols()
    ) %>% 
      mutate(gene_set = 'expressed genes'),
    read_tsv(
      here('data/spatial/salt_ma_summarised_perrineau_cis.tsv'),
      col_types = cols()
    ) %>% 
      mutate(gene_set = 'salt genes')
  )

salt_ma_final_summarised %>% 
  ggplot(aes(x = annotation, y = prop, fill = sample_factor)) +
  geom_bar(
    stat = 'identity', 
    color = 'black',
    position = position_dodge(width = 0.9))  +
  geom_errorbar(
    aes(ymin = lower_ci, ymax = upper_ci, group = sample_factor),
    position = position_dodge(width = 0.9),
    size = 0.5, width = 0.25) +
  facet_grid(
    ~ gene_set, scales = 'free') +
  theme_classic() +
  theme(
    axis.text = element_text(size = 12, color = 'black'),
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.title = element_text(size = 12, color = 'black'),
    legend.title = element_blank(),
    # legend.position = c(0.2, 0.88),
    legend.position = 'top',
    legend.background = element_blank(),
    legend.text = element_text(size = 10, color = 'black')
  ) +
  coord_cartesian(
    y = c(0, 1)
  ) +
  scale_x_discrete(
    limits = c('CDS', 'utr5', 'utr3', 'intronic', 'intergenic')
  ) +
  scale_fill_manual(
    values = palette[c(6, 7)]
  )

```

Fig 2 combined - featuring MA, HS, salt 0, and salt 5:

```{r}
fig_2_all = ma_orig_final %>% 
  bind_rows(adaptation_final) %>% 
  bind_rows(salt_ma_final_summarised) %>% 
  mutate(
    gene_set = factor(
      gene_set, levels = c('genome wide', 'expressed genes', 'salt genes')),
    annotation = factor(
      annotation, levels = c('CDS', 'utr5', 'utr3', 'intronic', 'intergenic')),
    sample_factor = case_when(
      sample_factor == 'ma_original' ~ 'MA',
      sample_factor == 'adaptation' ~ 'HS',
      sample_factor == 'salt 0' ~ 'Salt MA (control)',
      sample_factor == 'salt 5' ~ 'Salt MA (salt)'
    ),
    sample_factor = factor(
      sample_factor, 
      levels = c('MA', 'HS', 'Salt MA (control)', 'Salt MA (salt)')),
  )

# write_tsv(fig_2_all, here('data/plots/fig_2_all.tsv'))
```

```{r}
fig_2_all = read_tsv(here('data/plots/fig_2_all.tsv'))

fig_2_all %>% 
 ggplot(
    aes(x = annotation, y = prop, fill = sample_factor)) +
  geom_bar(
    stat = 'identity', 
    position = position_dodge(width = 0.9),
    colour = 'black',
    size = 0.4) +
  geom_errorbar(
    aes(ymin = lower_ci, ymax = upper_ci),
    position = position_dodge(width = 0.9),
    size = 0.4, width = 0.2
  ) +
  facet_grid(
    ~ gene_set, scales = 'free') +
  coord_cartesian(
    y = c(0, 1)
  ) +
  labs(
    x = '',
    y = 'Proportion of all mutations'
  ) +
  scale_fill_manual(
    values = c(
      'MA' = '#87C5FC', 'HS' = '#FF5B5F', 
      'Salt MA (control)' = palette[6], 'Salt MA (salt)' = palette[7])
  ) +
  # macro elements
  guides(fill = guide_legend(
    override.aes = list(color = 'white'))
  ) +
  theme(
    panel.border = element_rect(colour = 'black', size = 0.5, fill = NA),
    panel.background = element_blank(),
    strip.background = element_blank(),
    legend.position = c(0.175, 0.8),
    legend.text = element_text(size = 8, colour = 'black'),
    legend.title = element_blank(),
    legend.key = element_rect(color = 'white'),
    legend.key.size = unit(0.5, 'cm')) +
  # text elements
  theme(
    strip.text = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.title = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.text = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )
```


