---
title: "Intermutation distances"
author: Ahmed Hasan
output: pdf_notebook
---

# Import packages

```{r}
library(tidyverse)
library(fs)
library(here)
```

# Load in and clean data

```{r}
muts_raw = read_tsv(
  here('data/mutations/mut_describer/muts_described.final.tsv'),
  col_types = cols()
)

muts_clean = muts_raw %>% 
  
  # remove extra columns
  select(chromosome, position, mutant_sample) %>% 
  
  # add sample and line groups as new cols
  mutate(
    sample_group = str_extract(mutant_sample, '^[A-Z]{2}'),
    line = str_extract(mutant_sample, '[05]$')
  ) %>% 
  
  # sort
  mutate(chrom_val = as.numeric(str_extract(chromosome, '[0-9]{1,2}$'))) %>% 
  arrange(line, chrom_val, sample_group, position)

ma_original = read_tsv(
  here('data/spectrum_context/MA/final.curated_muts.coord_sorted.txt'),
  col_types = cols(), na = c('.')) %>% 
  select(-chromosome_1, -position_1)
  # filter(nchar(mutation) == 3) # remove indels

# for adaptation dataset - see mut_base_spectrum.Rmd
splice_sites = data.frame(
  chromosome = c('chromosome_12', 'chromosome_5', 'chromosome_6', 'chromosome_8'),
  position = c(8771346, 2315498, 2122297, 1131421)
)

adaptation = read_csv(
  here('data/prev/all_mutations_w_shared_hmmIBD_corrected_FPKM.csv'),
  # here('data/prev/all_mutations.csv'),
  col_types = cols(), na = c('.')
) %>% 
  # filter(type == 'SNP') %>% 
  # adding a silent/nonsilent column to the adaptation data frame
  mutate(site_type = case_when(
    Nonsyn_V_Syn == 'synonymous' | intronic == 1 | intergenic == 1 ~ 'silent',
    TRUE ~ 'nonsilent'
  )) %>% 
  mutate(site_type = ifelse(
    chromosome %in% splice_sites$chromosome & position %in% splice_sites$position, 'nonsilent', site_type
  )) # 119 each
```

# Distances

Currently just working off chromosomes - ignoring scaffolds and organelles

## Ungrouped

Intermutation distances across entire mut dataset - no grouping

```{r}
muts_clean %>% 
  arrange(chrom_val, position) %>% 
  group_by(chromosome) %>% 
  
  # get intermut distances
  mutate(next_pos = lead(position)) %>% 
  mutate(d = next_pos - position) %>% 
  
  # remove 'outer boundary' muts
  filter(!is.na(d))
```

## Grouped by line

Intermutation distances in 0 lines and 5 lines, across all sample groups

```{r}
muts_line = muts_clean %>% 
  arrange(line, chrom_val, position) %>% 
  group_by(line, chromosome) %>% 
  
  # get intermut distances
  mutate(next_pos = lead(position)) %>% 
  mutate(d = next_pos - position) %>% 
  
  # remove 'outer boundary' muts
  filter(!is.na(d))

muts_line
```

## Grouped by line and sample group

```{r}
muts_line_group = muts_clean %>% 
  arrange(line, sample_group, chrom_val, position) %>% 
  group_by(line, sample_group, chromosome) %>% 
  
  # get intermut distances
  mutate(next_pos = lead(position)) %>% 
  mutate(d = next_pos - position) %>% 
  
  # remove 'outer boundary' muts
  filter(!is.na(d)) %>% 
  ungroup()

muts_line_group
```

# Kolmogorov-Smirnov tests

## Plots

Across groups and lines:

```{r}
muts_line_group %>% 
  filter(sample_group != 'SL') %>%
  ggplot(aes(x = log10(d))) +
  stat_ecdf(aes(color = sample_group, linetype = line)) +
  theme_classic() +
  scale_colour_viridis_d(begin = 0.4, end = 0.6) +
  stat_ecdf(aes(color = sample_group, linetype = line))
```

## KS tests

### Between 0 and 5 of the same group

```{r}
ks.test(
  muts_line_group %>% filter(sample_group == 'CC', line == 0) %>% select(d) %>% unlist(),
  muts_line_group %>% filter(sample_group == 'CC', line == 5) %>% select(d) %>% unlist()
)

ks.test(
  muts_line_group %>% filter(sample_group == 'DL', line == 0) %>% select(d) %>% unlist(),
  muts_line_group %>% filter(sample_group == 'DL', line == 5) %>% select(d) %>% unlist()
)
```

### Between sample groups

```{r}
ks.test(
  muts_line_group %>% filter(sample_group == 'CC', line == 0) %>% select(d) %>% unlist(),
  muts_line_group %>% filter(sample_group == 'DL', line == 0) %>% select(d) %>% unlist()
)

ks.test(
  muts_line_group %>% filter(sample_group == 'CC', line == 5) %>% select(d) %>% unlist(),
  muts_line_group %>% filter(sample_group == 'DL', line == 5) %>% select(d) %>% unlist()
)
```

### Between salt MA and original MA

```{r}
# prepping original MA dataset
ma_orig_distances = ma_original %>% 
  select(chromosome, position, mutant_sample) %>% 
  mutate(chrom_val = as.numeric(str_extract(chromosome, '[0-9]{1,2}$'))) %>% 
  arrange(chrom_val, position) %>% 
  group_by(chromosome) %>% 
  
  # get intermut distances
  mutate(next_pos = lead(position)) %>% 
  # mutate(next_pos = ifelse(is.na(next_pos), lag(position), next_pos)) %>% 
  mutate(d = next_pos - position) %>% 
  
  # remove 'outer boundary' muts
  # filter(!is.na(d)) %>% 
  ungroup()

# plot
muts_line_group %>% 
  filter(sample_group != 'SL') %>%
  ggplot(aes(x = log10(d))) +
  stat_ecdf(aes(color = sample_group, linetype = line)) +
  stat_ecdf(data = ma_orig_distances) +
  theme_classic() +
  scale_colour_viridis_d(begin = 0.4, end = 0.8)
```

KS tests:

```{r}
ks.test(
  muts_line_group %>% filter(sample_group == 'CC', line == 0) %>% select(d) %>% unlist(),
  ma_orig_distances %>% select(d) %>% unlist()
)

ks.test(
  muts_line_group %>% filter(sample_group == 'CC', line == 5) %>% select(d) %>% unlist(),
  ma_orig_distances %>% select(d) %>% unlist()
)

ks.test(
  muts_line_group %>% filter(sample_group == 'DL', line == 0) %>% select(d) %>% unlist(),
  ma_orig_distances %>% select(d) %>% unlist()
)

ks.test(
  muts_line_group %>% filter(sample_group == 'DL', line == 5) %>% select(d) %>% unlist(),
  ma_orig_distances %>% select(d) %>% unlist()
)
```

## Adaptation dataset

```{r}
# prepping original MA dataset
adaptation_distances = adaptation %>% 
  split(.$site_type) %>% 
  map(
    ~ select(., chromosome, position, mutant_sample, site_type) %>% 
    mutate(chrom_val = as.numeric(str_extract(chromosome, '[0-9]{1,2}$'))) %>% 
    arrange(chrom_val, position) %>% 
    group_by(chromosome) %>% 
    
    # get intermut distances
    mutate(next_pos = lead(position)) %>% 
    mutate(d = next_pos - position) %>% 
    
    # remove 'outer boundary' muts
    filter(!is.na(d)) %>% 
    ungroup()
  )

adaptation_silent = adaptation_distances$silent
adaptation_nonsilent = adaptation_distances$nonsilent
```

Recreating the original Fig 4:

```{r}
ma_orig_distances %>% 
  mutate(site_type = 'ma') %>% 
  bind_rows(adaptation_silent, adaptation_nonsilent) %>% 

ggplot(aes(x = log10(d), color = site_type)) +
  stat_ecdf(size = 1.2, geom = 'line') +
  theme_classic() +
  scale_colour_viridis_d(begin = 0.4, end = 0.8)
```

```{r}
# between MA and silent HS sites
ks.test(
  ma_orig_distances$d, adaptation_silent$d
)

# between silent and nonsilent HS sites
ks.test(
  adaptation_silent$d, adaptation_nonsilent$d
)
```


















