---
title: "Intermutation distances"
author: Ahmed Hasan
output: pdf_notebook
---

# Import packages

```{r}
library(tidyverse)
library(fs)
library(here)
library(wesanderson)
library(patchwork)
library(broom)
```

# Load in and clean data

```{r}
muts_raw = read_tsv(
  here('data/mutations/mut_describer/muts_described.final.tsv'),
  col_types = cols()
) %>% 
  bind_rows(
    read_tsv(
      here('data/mutations/mut_describer/indels_described.tsv'),
      col_types = cols()
    ) %>% 
      mutate(mut_codon = as.character(mut_codon),
             rho = as.character(rho))
  )

muts_clean = muts_raw %>% 
  
  # remove extra columns
  select(chromosome, position, mutant_sample) %>% 
  
  # add sample and line groups as new cols
  mutate(
    sample_group = str_extract(mutant_sample, '^[A-Z]{2}'),
    line = str_extract(mutant_sample, '[05]$')
  ) %>% 
  
  # sort
  mutate(chrom_val = as.numeric(str_extract(chromosome, '[0-9]{1,2}$'))) %>% 
  arrange(line, chrom_val, sample_group, position) 

ma_original = read_tsv(
  here('data/spectrum_context/MA/final.curated_muts.coord_sorted.txt'),
  col_types = cols(), na = c('.')) %>% 
  select(-chromosome_1, -position_1)
  # filter(nchar(mutation) == 3) # remove indels

# for adaptation dataset - see mut_base_spectrum.Rmd
splice_sites = data.frame(
  chromosome = c('chromosome_12', 'chromosome_5', 'chromosome_6', 'chromosome_8'),
  position = c(8771346, 2315498, 2122297, 1131421)
)

adaptation = read_tsv(
  # here('data/prev/all_mutations_w_shared_hmmIBD_corrected_FPKM.csv'),
  # here('data/prev/all_mutations.csv'),
  here('data/rate/mut_describer/adaptation.gene_sets.filtered.tsv'),
  col_types = cols(), na = c('.')
) %>%
  select(-cluster) %>% 
  # filter(type == 'SNP') %>% 
  # adding a silent/nonsilent column to the adaptation data frame
  mutate(site_type = case_when(
    Nonsyn_V_Syn == 'synonymous' | intronic == 1 | intergenic == 1 ~ 'silent',
    TRUE ~ 'nonsilent'
  )) %>% 
  mutate(site_type = ifelse(
    chromosome %in% splice_sites$chromosome & position %in% splice_sites$position, 'nonsilent', site_type
  )) # 119 each
```

# Distances

Currently just working off chromosomes - ignoring scaffolds and organelles

## Ungrouped

Intermutation distances across entire mut dataset - no grouping

```{r}
muts_clean %>% 
  arrange(chrom_val, position) %>% 
  group_by(chromosome) %>% 
  
  # get intermut distances
  mutate(next_pos = lead(position)) %>% 
  mutate(d = next_pos - position) %>% 
  
  # remove 'outer boundary' muts
  filter(!is.na(d))
```

## Grouped by line

Intermutation distances in 0 lines and 5 lines, across all sample groups

```{r}
muts_line = muts_clean %>% 
  arrange(line, chrom_val, position) %>% 
  group_by(line, chromosome) %>% 
  
  # get intermut distances
  mutate(next_pos = lead(position)) %>% 
  mutate(d = next_pos - position) %>% 
  
  # remove 'outer boundary' muts
  filter(!is.na(d)) %>% 
  ungroup()

muts_line
```

## Grouped by line and sample group

```{r}
muts_line_group = muts_clean %>% 
  arrange(line, sample_group, chrom_val, position) %>% 
  group_by(line, sample_group, chromosome) %>% 
  
  # get intermut distances
  mutate(next_pos = lead(position)) %>% 
  mutate(d = next_pos - position) %>% 
  
  # remove 'outer boundary' muts
  filter(!is.na(d)) %>% 
  ungroup()

muts_line_group

# write_tsv(muts_line_group, here('data/spatial/saltMA_intermut.tsv'))
```

# Kolmogorov-Smirnov tests

## Plots

Across lines - 

```{r}
muts_line %>% 
  ggplot(aes(x = log10(d))) +
  stat_ecdf(aes(color = line)) +
  theme_classic() +
  scale_colour_viridis_d(begin = 0.4, end = 0.6)
```


Across groups and lines:

```{r}
muts_line_group %>% 
  filter(sample_group != 'SL') %>%
  ggplot(aes(x = log10(d))) +
  stat_ecdf(aes(color = sample_group, linetype = line)) +
  theme_classic() +
  scale_colour_viridis_d(begin = 0.4, end = 0.6)
```

## KS tests

### Between 0 and 5

```{r}
ks.test(
  muts_line %>% filter(line == 0) %>% select(d) %>% unlist(),
  muts_line %>% filter(line == 5) %>% select(d) %>% unlist()
)
```


### Between 0 and 5 of the same group

```{r}
ks.test(
  muts_line_group %>% filter(sample_group == 'CC', line == 0) %>% select(d) %>% unlist(),
  muts_line_group %>% filter(sample_group == 'CC', line == 5) %>% select(d) %>% unlist()
)

ks.test(
  muts_line_group %>% filter(sample_group == 'DL', line == 0) %>% select(d) %>% unlist(),
  muts_line_group %>% filter(sample_group == 'DL', line == 5) %>% select(d) %>% unlist()
)
```

### Between sample groups

```{r}
# both of these are significant!
ks.test(
  muts_line_group %>% filter(sample_group == 'CC', line == 0) %>% select(d) %>% unlist(),
  muts_line_group %>% filter(sample_group == 'DL', line == 0) %>% select(d) %>% unlist()
) 

ks.test(
  muts_line_group %>% filter(sample_group == 'CC', line == 5) %>% select(d) %>% unlist(),
  muts_line_group %>% filter(sample_group == 'DL', line == 5) %>% select(d) %>% unlist()
)
```

### Between salt MA and original MA

```{r}
# prepping original MA dataset
ma_orig_distances = ma_original %>% 
  select(chromosome, position, mutant_sample) %>% 
  mutate(chrom_val = as.numeric(str_extract(chromosome, '[0-9]{1,2}$'))) %>% 
  arrange(chrom_val, position) %>% 
  distinct() %>% 
  group_by(chromosome) %>% 
  
  # get intermut distances
  mutate(next_pos = lead(position),
         prev_pos = lag(position)) %>% 
  # mutate(next_pos = ifelse(is.na(next_pos), lag(position), next_pos)) %>% 
  mutate(d = next_pos - position) %>% 
  mutate(d = ifelse(is.na(d), position - prev_pos, d)) %>% 
  
  # remove 'outer boundary' muts
  # filter(!is.na(d)) %>% 
  ungroup()

# plot
muts_line %>% 
  ggplot(aes(x = log10(d))) +
  stat_ecdf(aes(color = line)) +
  stat_ecdf(data = ma_orig_distances) +
  theme_classic() +
  scale_colour_viridis_d(begin = 0.4, end = 0.8)
```

KS tests:

```{r}
ks.test(
  muts_line %>% filter(line == 0) %>% select(d) %>% unlist(),
  ma_orig_distances %>% select(d) %>% unlist()
)

ks.test(
  muts_line %>% filter(line == 5) %>% select(d) %>% unlist(),
  ma_orig_distances %>% select(d) %>% unlist()
)

ks.test(
  muts_line_group %>% filter(sample_group == 'CC', line == 0) %>% select(d) %>% unlist(),
  ma_orig_distances %>% select(d) %>% unlist()
)

ks.test(
  muts_line_group %>% filter(sample_group == 'CC', line == 5) %>% select(d) %>% unlist(),
  ma_orig_distances %>% select(d) %>% unlist()
)

ks.test(
  muts_line_group %>% filter(sample_group == 'DL', line == 0) %>% select(d) %>% unlist(),
  ma_orig_distances %>% select(d) %>% unlist()
)

ks.test(
  muts_line_group %>% filter(sample_group == 'DL', line == 5) %>% select(d) %>% unlist(),
  ma_orig_distances %>% select(d) %>% unlist()
)
```

## Adaptation dataset

```{r}
adaptation_distances = adaptation %>% 
  split(.$site_type) %>% 
  map(
    ~ select(., chromosome, position, mutant_sample, site_type) %>% 
    mutate(chrom_val = as.numeric(str_extract(chromosome, '[0-9]{1,2}$'))) %>% 
    arrange(chrom_val, position) %>% 
    group_by(chromosome) %>% 
    
    # get intermut distances
    mutate(next_pos = lead(position)) %>% 
    mutate(d = next_pos - position) %>% 
    
    # remove 'outer boundary' muts
    filter(!is.na(d)) %>% 
    ungroup()
  )

adaptation_silent = adaptation_distances$silent
adaptation_nonsilent = adaptation_distances$nonsilent

# combined
adaptation_combined_distances = adaptation %>% 
    select(chromosome, position, mutant_sample, site_type) %>% 
    mutate(chrom_val = as.numeric(str_extract(chromosome, '[0-9]{1,2}$'))) %>% 
    arrange(chrom_val, position) %>% 
    group_by(chromosome) %>% 
    
    # get intermut distances
    mutate(next_pos = lead(position),
           prev_pos = lag(position)) %>% 
    mutate(d = next_pos - position) %>% 
    mutate(d = ifelse(is.na(d), position - prev_pos, d)) %>% # deal with 'outer boundary' muts
    
    ungroup()
  
```

Recreating the original Fig 4:

```{r}
ma_orig_distances %>% 
  mutate(site_type = 'ma') %>% 
  bind_rows(adaptation_silent, adaptation_nonsilent) %>% 

ggplot(aes(x = log10(d), color = site_type)) +
  stat_ecdf(size = 1.2, geom = 'line') +
  theme_classic() +
  scale_colour_viridis_d(begin = 0.4, end = 0.8)
```

```{r}
# between MA and silent HS sites
ks.test(
  ma_orig_distances$d, adaptation_silent$d
)

# between silent and nonsilent HS sites
ks.test(
  adaptation_silent$d, adaptation_nonsilent$d
)

# between salt MA and adaptation silent
ks.test(
  muts_line %>% filter(line == 0) %>% select(d) %>% unlist(),
  adaptation_silent$d
)

ks.test(
  muts_line %>% filter(line == 5) %>% select(d) %>% unlist(),
  adaptation_silent$d
)

# between salt MA and adaptation nonsilent
ks.test(
  muts_line %>% filter(line == 0) %>% select(d) %>% unlist(),
  adaptation_nonsilent$d
)

ks.test(
  muts_line %>% filter(line == 5) %>% select(d) %>% unlist(),
  adaptation_nonsilent$d
)
```

Comparisons with the full adaptation dataset:

```{r}
ks.test(
  muts_line %>% filter(line == 0) %>% select(d) %>% unlist(),
  adaptation_combined_distances$d
)

ks.test(
  muts_line %>% filter(line == 5) %>% select(d) %>% unlist(),
  adaptation_combined_distances$d
)

ks.test(
  ma_orig_distances$d, adaptation_combined_distances$d
)

ks.test(
  ma_orig_distances$d, 
  muts_line %>% filter(line == 5) %>% select(d) %>% unlist()
)

```

### Comparisons after downsampling to the adaptation dataset

```{r}
nrow(adaptation_combined_distances) # 238 mutations
nrow(muts_line %>% filter(line == 5)) # 174 mutations
nrow(ma_orig_distances) # 6798 muts
```

Need to downsample the other two mutation sets to 174 (salt 5 mut count)

```{r}
resample_ks_test = function(d1, d2, min_size = FALSE, verbose = TRUE) {
  if (!min_size) {
    n_muts = min(nrow(d1), nrow(d2))
  } else {
    n_muts = min_size
  }
  if (verbose == TRUE) {
    message(paste('sampling down to', n_muts))
  }
  d1_sampled = d1 %>% 
    slice_sample(n = n_muts, replace = FALSE) %>% 
    arrange(chrom_val, position) %>% 
    distinct() %>% 
    group_by(chromosome) %>% 
    mutate(next_pos = lead(position), prev_pos = lag(position),
           d = next_pos - position) %>% 
    mutate(d = ifelse(is.na(d), position - prev_pos, d))
  d2_sampled = d2 %>% 
    slice_sample(n = n_muts, replace = FALSE) %>% 
    arrange(chrom_val, position) %>% 
    distinct() %>% 
    group_by(chromosome) %>% 
    mutate(next_pos = lead(position), prev_pos = lag(position),
           d = next_pos - position) %>% 
    mutate(d = ifelse(is.na(d), position - prev_pos, d))
    
  out = broom::tidy(
    suppressWarnings(ks.test(d1_sampled$d, d2_sampled$d))
  )
  return(out)
}

resample_ks_test(
  ma_orig_distances,
  adaptation_combined_distances)
```

```{r}
# repeat 100 times across the three datasets
resampled_ks = list()

# MA vs HS
resampled_ks$ma_hs = map_dfr(
  c(1:100),
  ~ resample_ks_test(
    ma_orig_distances,
    adaptation_combined_distances,
    min_size = 174,
    verbose = FALSE
  )
)

# MA vs salt MA
resampled_ks$ma_salt = map_dfr(
  c(1:100),
  ~ resample_ks_test(
    ma_orig_distances,
    muts_line %>% filter(line == 5),
    verbose = FALSE
  )
)

# HS vs salt MA
resampled_ks$hs_salt = map_dfr(
  c(1:100),
  ~ resample_ks_test(
    adaptation_combined_distances,
    muts_line %>% filter(line == 5),
    verbose = FALSE
  )
)
```

Looking at the distribution of results:

```{r}
summary(resampled_ks$ma_hs %>% select(statistic, p.value))

summary(resampled_ks$ma_salt %>% select(statistic, p.value))

summary(resampled_ks$hs_salt %>% select(statistic, p.value))
```




# Final combined plot

Creating a downsampled dataset:

```{r}
set.seed(42)
downsample_size = nrow(muts_line %>% filter(line == 5)) # 174

final_distances_downsampled = muts_line %>% 
  rename(site_type = line) %>% 
  filter(site_type != '0') %>% 
  select(
    chromosome, position, mutant_sample,
    chrom_val, next_pos, d, site_type
  ) %>% 
  bind_rows(
    ma_orig_distances %>% 
      slice_sample(n = downsample_size, replace = FALSE) %>% 
      arrange(chrom_val, position) %>% 
      distinct() %>% 
      group_by(chromosome) %>% 
      mutate(next_pos = lead(position), prev_pos = lag(position),
             d = next_pos - position) %>% 
      mutate(d = ifelse(is.na(d), position - prev_pos, d)) %>% 
      mutate(site_type = 'ma'),
    adaptation_combined_distances %>% 
      slice_sample(n = downsample_size, replace = FALSE) %>% 
      arrange(chrom_val, position) %>% 
      distinct() %>% 
      group_by(chromosome) %>% 
      mutate(next_pos = lead(position), prev_pos = lag(position),
             d = next_pos - position) %>% 
      mutate(d = ifelse(is.na(d), position - prev_pos, d)) %>% 
      mutate(site_type = 'adaptation')
) %>% 
   mutate(
    site_type = case_when(
      site_type == 'adaptation' ~ 'HS',
      site_type == 'ma' ~ 'MA',
      site_type == '5' ~ 'Salt MA'
    )) %>% 
  mutate(
    site_type = fct_relevel(site_type, 
      c('MA', 'HS', 'Salt MA'))
  )

# final_distances = read_tsv(here('data/plots/intermutation_distances.tsv'), col_types = cols())


# final_distances_downsampled = final_distances %>% 
#   split(.$site_type) %>% 
#   map_dfr(~ slice_sample(., n = downsample_size, replace = FALSE))
```

Plotting:

```{r}
distance_cdf_plot = final_distances_downsampled %>% 
  ggplot(
    aes(x = log10(d), color = site_type)) +
  stat_ecdf(
    size = 1, 
    geom = 'line') +
  scale_color_manual(
    values = wes_palette(
      5, name = 'Darjeeling1', type = 'continuous')[c(1, 3, 5)]) +
  labs(
    x = '',
    y = 'Cumulative density'
  ) +
  # legend styling
  theme(
    legend.text = element_text(family = 'Helvetica', color = 'black', size = 10),
    legend.position = c(0.15, 0.8),
    legend.title = element_blank(),
    legend.box.background = element_blank(),
    legend.key = element_blank()
  ) +
  theme(
    panel.background = element_blank(),
    axis.line = element_line(size = 0.5, color = 'black'),
    axis.ticks = element_line(size = 0.5, color = 'black'),
    axis.title = element_text(family = 'Helvetica', size = 12, color = 'black'),
    axis.text = element_text(family = 'Helvetica', size = 12, color = 'black'),
    axis.text.x = element_blank() # x axis will be in lower facet
  )

distance_cdf_plot
```

Creating the density plot:

```{r}
final_distances_density = final_distances_downsampled %>% 
  mutate(
    log_d = log10(d),
    bin = floor(log_d / 0.5) * 0.5) %>% 
  # get density vals
  group_by(
    site_type, bin
  ) %>% 
  summarise(
    n = n()
  ) %>% 
  filter( # remove end of chr mutations
    !is.na(bin), 
    is.finite(bin)) %>% 
  mutate(
    total_group = sum(n)) %>% 
  ungroup() %>% 
  mutate(
    density = n / total_group) 

# add missing values
final_distances_missing = final_distances_density %>% 
  split(.$site_type) %>% 
  map_dfr(
    function(d) {
      bin_range = seq(0, 6.5, 0.5)
      current_vals = unique(d$bin)
      missing_vals = bin_range[
        sapply(bin_range, function(val) !val %in% current_vals)
      ]
      if (length(missing_vals) > 0) {
        vals_to_add = data.frame(
          bin = missing_vals
        )
        vals_to_add$site_type = unique(d$site_type)
        vals_to_add$bin = missing_vals
        vals_to_add$n = 0
        vals_to_add$total_group = unique(d$total_group)
        vals_to_add$density = 0
        vals_to_add = select(vals_to_add, site_type, bin, everything())
        return(vals_to_add)
      }
    }
  )

final_distances_density = bind_rows(
  final_distances_density,
  final_distances_missing
) %>% 
  arrange(site_type, bin)

# write_tsv(final_distances, here('data/plots/intermutation_distances.tsv'))
# write_tsv(final_distances_density, here('data/plots/intermutation_density.tsv'))
```
  
Density barplot:

```{r}
# final_distances_density = read_tsv(here('data/plots/intermutation_density.tsv'), col_types = cols())

distance_density_plot = final_distances_density %>% 
  ggplot(
    aes(x = bin, y = density, fill = site_type)
  ) +
  geom_bar(
    stat = 'identity', 
    position = position_dodge(),
    color = 'black'
  ) +
  guides(fill = FALSE) +
  scale_fill_manual(
    values = wes_palette(
      5, name = 'Darjeeling1', type = 'continuous')[c(1, 3, 5)]) +
  labs(
    x = expression(
      paste('Intermutation distance (', log[10], ' bp)')),
    y = 'Proportion'
  ) +
  # legend styling
  theme(
    panel.background = element_blank(),
    axis.line = element_line(size = 0.5, color = 'black'),
    axis.ticks = element_line(size = 0.5, color = 'black'),
    axis.title = element_text(family = 'Helvetica', size = 12, color = 'black'),
    axis.text = element_text(family = 'Helvetica', size = 12, color = 'black')
  )

distance_density_plot
  
```

Combining the two:

```{r}
combined_fig = distance_cdf_plot + distance_density_plot +
  plot_layout(
    nrow = 2, ncol = 1,
    heights = c(3, 2)
  )

combined_fig

ggsave(here('plots/intermutation_distances.eps'), plot = combined_fig, width = 7.5, height = 5)
```

# Mutation clusters

```{r}
final_distances %>% 
  filter(site_type == 'HS lines') %>% 
    # get mutations that are part of clusters
  mutate(prev_pos = lag(position),
         prev_d = abs(position - prev_pos),
         prev_chrom = lag(chrom_val))
  filter(d <= 10 | prev_d <= 10)
```


```{r}
label_clusters = function(d) {
  d_labelled = d %>% 
    # get mutations that are part of clusters
    mutate(prev_pos = lag(position),
           prev_d = abs(position - prev_pos),
           prev_chrom = lag(chrom_val)) %>% 
    filter(d <= 10 | prev_d <= 10) %>% 
    mutate(next_pos = lead(position),
           prev_pos = lag(position)) %>%  # overwrite prev next_pos column
    
    # label clusters
    mutate(prev_mut = ifelse(
      abs(position - next_pos) > 10, TRUE, FALSE)) %>% 
    mutate(prev_mut = ifelse(
      abs(position - prev_pos) > 10, TRUE, FALSE) # gets 'right bound mut' for each cluster
    ) %>% 
    mutate(prev_mut = ifelse(is.na(prev_mut), FALSE, prev_mut)) %>% # just for first row
    mutate(
      cluster = cumsum(prev_mut) # uses bool var assigned above to assign clusters - increments at each TRUE
    )
  return(d_labelled)
}
```

Creating a list of clusters:

```{r}
clusters = list()

clusters = final_distances %>% 
  split(.$site_type) %>% 
  map(~ label_clusters(.))

names(clusters) = c('ma', 'adaptation', 'salt_ma')

```

Getting the number of clusters for each:

```{r}
# add 1! clusters are 0-indexed
clusters %>% 
  map(~ max(.$cluster))
```

Two proportions z-test:

```{r}
prop.test(
  x = c(15, 155), 
  n = c(nrow(adaptation_combined_distances), nrow(ma_orig_distances)))

prop.test(
  x = c(15, 4),
  n = c(nrow(adaptation_combined_distances), nrow(final_distances %>% filter(site_type == 'Salt MA')))
)

prop.test(
  x = c(4, 155),
  n = c(nrow(final_distances %>% filter(site_type == 'Salt MA')), nrow(ma_orig_distances))
)
```

Getting the annotations of the HS clusters:

```{r}
clusters$adaptation %>% 
  select(chromosome, position, cluster) %>% 
  
  left_join(adaptation, by = c('chromosome', 'position')) %>% 
  
  # clean up
  select(chromosome, position, cluster, CDS, intronic:utr3) %>% 
  
  # sum annotations
  group_by(cluster) %>% 
  summarise(across(.cols = CDS:utr3, .fns = sum)) %>% 
  
  # transform
  pivot_longer(cols = CDS:utr3, names_to = 'annotation') %>% 
  filter(value != 0) %>% 
  group_by(cluster, annotation) %>% # give each cluster a single annotation
  count() %>% 
  
  # get final per-annotation counts
  group_by(annotation) %>% 
  count()
```

How many clusters are in candidate salt genes?

```{r}
clusters$adaptation %>% 
  select(chromosome, position, cluster) %>% 
  
  left_join(adaptation, by = c('chromosome', 'position')) %>% 
  select(chromosome, position, cluster, Gene.primaryIdentifier, contains('perrineau')) %>% 
  
  filter(perrineau == 1)

# 3 of 14 clusters are in perrineau genes 
```