---
title: "Mutation base spectrum"
author: Ahmed Hasan
output: pdf_document
---

# Package import

```{r}
library(tidyverse)
library(fs)
library(here)
library(wesanderson)
```

# Data import

```{r}
# all mut describer outputs
salt <- read_tsv(
  here('data/rate/mut_describer/muts_described.final.gene_sets.tsv'), col_types = cols()) %>% 
  mutate(sample_type = str_extract(mutant_sample, '[05]$'),
         sample_group = str_extract(mutant_sample, '^[A-Z]{2}')) %>% 
  mutate(sample_factor = paste(sample_group, sample_type, sep = '_'))

ma <- read_tsv(
  here('data/spectrum_context/MA/final.curated_muts.coord_sorted.txt'), 
  col_types = cols(), na = c('.')
) %>% 
  select(-chromosome_1, -position_1) %>% 
  filter(nchar(mutation) == 3) # remove indels

adaptation <- read_csv(
  here('data/prev/all_mutations_w_shared_hmmIBD_corrected_FPKM.csv'), 
  col_types = cols(), na = c('.')
) %>% 
  filter(type == 'SNP')

# fxn to collapse callables - e.g. A and T should just be A
collapse_callables = function(d) {
  d_out = d %>% 
    mutate(
      A_count = A_count + T_count,
      C_count = C_count + G_count
    ) %>% 
    select(-T_count, -G_count)
  return(d_out)
}

# callables files
base_callables = read_tsv(
  here('data/rate/base_callables.tsv'),
  col_types = cols()
) %>% 
  collapse_callables()

ma_callables = read_tsv(
  here('data/rate/MA_base_callables.tsv'),
  col_types = cols()
) %>% 
  collapse_callables()

hs_callables = read_tsv(
  here('data/rate/HS_base_callables.tsv'),
  col_types = cols()
) %>% 
  collapse_callables()

hs_callables_silent = read_tsv(
  here('data/rate/HS_base_callables_silent.tsv'),
  col_types = cols()
) %>% 
  collapse_callables()

generations = read_tsv(
  here('data/rate/saltMA_SNM_rate_final.tsv'), 
  col_types = cols()
  ) %>% 
  mutate(
    sample_group = str_extract(sample, '^[A-Z]{2}'),
    sample_type = str_extract(sample, '[05]$'),
    sample_factor = paste(sample_group, sample_type, sep = '_')
  ) %>% 
  mutate(
    type = 'SNM'
  ) %>% 
  filter(sample_type == '5') %>% # only keep salt 5
  select(sample_factor, generations) %>% 
  summarise(mean_generations = mean(generations)) # 236

```

# Mutational base spectrum

6 possible SNMs
- A -> C (also T -> G)
- A -> G (also T -> C)
- A -> T (also T -> A)
- C -> A (also G -> T)
- C -> G (also G -> C)
- C -> T (also G -> A)

## Generating observed count tables

Need to get splice sites for silent/nonsilent classification of intronic sites - 
ran this on hpcnode1 using a gffutils db:

```{python, eval=FALSE}
import csv
import gffutils
from tqdm import tqdm

with open('muts.csv', 'r') as f: # equivalent of adaptation data frame above
    reader = csv.DictReader(f)
    reclist = [rec for rec in reader if rec['intronic'] == '1']
  
db = gffutils.create_db('final.strict.GFF3', dbfn='final.strict.db',
    force=True, keep_order=True, merge_strategy='merge', sort_attribute_values=True)
db = gffutils.FeatureDB('final.strict.db', keep_order=True)

splice_list = []
for record in tqdm(reclist):
    gene_name = record['Gene.primaryIdentifier']
    db_gene = db[gene_name]
    pos = int(record['position'])
    for exon in db.children(db_gene, featuretype='exon', order_by='start'):
        if min(abs(pos - exon.start), abs(pos - exon.end)) < 3:
        print(gene_name, record['chromosome'], record['position'])
        print(exon)
        splice_list.append(record)
```

Got me four intronic mutations that shouldn't be classed as silent:

```
Cre12.g545550 chromosome_12 8771346
chromosome_12	phytozome8_0	exon	8770962	8771344	.	+	.	ID=PAC:26907314.exon.5;pacid=26907314;Parent=PAC:26907314;ness_ID=26907314
g5391 chromosome_5 2315498
chromosome_5	phytozome8_0	exon	2315470	2315497	.	-	.	ID=PAC:26897662.exon.6;pacid=26897662;Parent=PAC:26897662;ness_ID=26897662
Cre06.g265550 chromosome_6 2122297
chromosome_6	phytozome8_0	exon	2122193	2122295	.	-	.	ID=PAC:26893506.exon.4;pacid=26893506;Parent=PAC:26893506;ness_ID=26893506
Cre08.g362650 chromosome_8 1131421
chromosome_8	phytozome8_0	exon	1131423	1131551	.	-	.	ID=PAC:26888200.exon.20;pacid=26888200;Parent=PAC:26888200;ness_ID=26888200
```

Creating a data frame for these:

```{r}
splice_sites = data.frame(
  chromosome = c('chromosome_12', 'chromosome_5', 'chromosome_6', 'chromosome_8'),
  position = c(8771346, 2315498, 2122297, 1131421)
)

# adding a silent/nonsilent column to the adaptation data frame
adaptation_silent_nonsilent = adaptation %>% 
  filter(utr5 == 0, utr3 == 0) %>% 
  mutate(site_type = case_when(
    Nonsyn_V_Syn != 'synonymous' & intronic == 0 & intergenic == 0 ~ 'nonsilent',
    Nonsyn_V_Syn == 'synonymous' | intronic == 1 | intergenic == 1 ~ 'silent'
  )) %>% 
  mutate(site_type = ifelse(
    chromosome %in% splice_sites$chromosome & position %in% splice_sites$position, 'nonsilent', site_type
  )) # 119 each

```

Getting mutation counts:

```{r}
get_counts = function(d) {
  count_table <- d %>% 
    select(mutation) %>% 
    mutate(
      type = case_when(
        mutation == 'T>G' ~ 'A>C',
        mutation == 'T>C' ~ 'A>G',
        mutation == 'T>A' ~ 'A>T',
        mutation == 'G>T' ~ 'C>A',
        mutation == 'G>C' ~ 'C>G',
        mutation == 'G>A' ~ 'C>T',
        TRUE ~ mutation
      )
    ) %>% 
    count(type, name = 'count')
  return(count_table)
}

count_table = list(
    # salt_nonsilent = salt %>% filter(nonsyn_v_syn != 'nonsynonymous'), 
    # salt_silent = salt %>% filter(nonsyn_v_syn == 'synonymous'),
    salt_0 = salt %>% filter(sample_type == '0', mutant_sample != 'DL51_0'),
    salt_5 = salt %>% filter(sample_type == '5', mutant_sample != 'DL51_5'),
    ma = ma, 
    adaptation = adaptation
    # adaptation_nonsilent = adaptation_silent_nonsilent %>% filter(site_type == 'nonsilent'),
    # adaptation_silent = adaptation_silent_nonsilent %>% filter(site_type == 'silent')
  ) %>% 
  map_dfr(., get_counts, .id = 'name')

# separating three salt MA groups in separate count table
salt_group_table = salt %>% 
  mutate(group_type = paste(sample_group, sample_type, sep = '_')) %>% 
  split(.$group_type) %>% 
  map_dfr(., get_counts, .id = 'name')
  
count_table

```

Summary stats:

```{r}
# most and least common mutations
count_table %>% 
  
  # get most and least mutations
  group_by(name) %>% 
  filter(count == max(count) | count == min(count)) %>% 
  mutate(mut_type = case_when(
    count == min(count) ~ 'least mutated',
    count == max(count) ~ 'most mutated')
    ) %>% 
  
  # get fold difference 
  mutate(fold_diff = max(count) / min(count))


```


## Calculate expected mutation counts

Expected mutation rates:

```{r}
obs_exp = count_table %>% 
  # get the total mutations per anc mut per sample (e.g. all A anc muts in saltMA)
  mutate(anc_mut = substr(type, 1, 1)) %>% 
  mutate(group = str_extract(name, '(salt_[05]|ma|adaptation)')) %>% 
  group_by(name) %>% 
  mutate(total_mut_anc = sum(count)) %>% 
  ungroup() %>% 
  # get expected number of muts (1/3 prob for either base)
  mutate(expected_count = case_when(
    anc_mut == 'A' ~ ceiling((total_mut_anc * 0.365) / 3),
    anc_mut == 'C' ~ ceiling((total_mut_anc * 0.635) / 3))
    ) %>% 
  # get obs / exp
  mutate(obs_exp = count / expected_count)

obs_exp

obs_exp_salt = salt_group_table %>% 
  mutate(anc_mut = substr(type, 1, 1)) %>% 
  group_by(name) %>% 
  mutate(total_mut_anc = sum(count)) %>% 
  ungroup() %>% 
  # get expected number of muts
  mutate(expected_count = case_when(
    anc_mut == 'A' ~ ceiling((total_mut_anc * 0.365) / 3),
    anc_mut == 'C' ~ ceiling((total_mut_anc * 0.635) / 3))
    ) %>% 
  # get obs / exp
  mutate(obs_exp = count / expected_count) %>% 
  # add missing rows
  add_row(name = 'CC_0', type = 'A>T') %>% 
  add_row(name = 'SL_0', type = 'C>G') %>% 
  add_row(name = 'SL_0', type = 'C>T') %>% 
  add_row(name = 'SL_5', type = 'A>T') %>% 
  add_row(name = 'SL_5', type = 'A>G') %>% 
  add_row(name = 'SL_5', type = 'C>T') %>% 
  add_row(name = 'SL_5', type = 'C>G') %>% 
  mutate(across(where(is.numeric), .fns = ~ ifelse(is.na(.), 0, .))) %>% 
  mutate(anc_mut = substr(type, 1, 1))

obs_exp_salt
```

## Calculate mutation spectrum relative to callables

Salt MA:

```{r}
base_callables_long = base_callables %>% 
  
  # summarise by sample factor
  mutate(
    sample_group = str_extract(sample, '^[A-Z]{2}'),
    sample_type = str_extract(sample, '[05]$'),
    sample_factor = paste(sample_group, sample_type, sep = '_')
  ) %>% 
  group_by(sample_factor) %>% 
  summarise(
    across(where(is.numeric), sum)
  ) %>% 
  
  # pivot to long format for join
  pivot_longer(
    cols = where(is.numeric), 
    names_to = 'base', values_to = 'callables'
  ) %>% 
  mutate(
    base = str_extract(base, '^[A-Z]')
  )

# nts: these callable values have the mutations 'baked in'
# e.g. if there were 10 C mutations in CC_0, then the C callables are inflated by 10
# unlikely to hugely sway the results, but may need to correct for this down the line

# join to mutation counts
salt_group_rate = salt_group_table %>% 
  rename(sample_factor = name) %>% 
  mutate(
    base = str_extract(type, '^[A-Z]{1}')
  ) %>% 
  
  # join
  left_join(
    base_callables_long, by = c('sample_factor', 'base')
  ) %>% 
  
  # get rate
  mutate(
    rate = count / callables
  )

salt_group_rate

# write_tsv(salt_group_rate, here('data/spectrum_context/saltMA_spectrum.tsv'))
```

Summarised over 0 and 5:

```{r}
base_callables_summarised = base_callables_long %>% 
  mutate(
    sample_type = str_extract(sample_factor, '[05]$')
  ) %>% 
  group_by(base, sample_type) %>% 
  summarise(callables = sum(callables))
  
salt_rate_summarised = salt_group_table %>% 
  mutate(
    base = str_extract(type, '^[AGCT]{1}'),
    sample_type = str_extract(name, '[05]$')
  ) %>% 
  group_by(sample_type, base, type) %>% 
  summarise(count = sum(count), .groups = 'drop') %>% 
  left_join(
    base_callables_summarised,
    by = c('sample_type', 'base')) %>% 
  mutate(rate = count / (callables * 236)) # sample count is baked in

salt_rate_summarised
```

Original MA and adaptation:

```{r}
transform_callables = function(d, sample_factor_name) {
  d %>% 
    
    mutate(sample_factor = sample_factor_name) %>% 
    group_by(sample_factor) %>% 
    summarise(
      across(where(is.numeric), sum)
    ) %>% 
    
    # pivot to long format for join
    pivot_longer(
      cols = where(is.numeric), 
      names_to = 'base', values_to = 'callables'
    ) %>% 
    mutate(
      base = str_extract(base, '^[A-Z]')
    )
}

ma_callables_long = transform_callables(ma_callables, 'ma')
hs_callables_long = transform_callables(hs_callables, 'adaptation') # this is combined

# hs_callables_long_silent = transform_callables(
  # hs_callables_silent %>% filter(type == 'silent'), 'adaptation_silent')
# hs_callables_long_nonsilent = transform_callables(
  # hs_callables_silent %>% filter(type == 'nonsilent'), 'adaptation_nonsilent')


```

Combining the mut counts with callables:

```{r}
rate_summarised = list()

make_final_count_table = function(d, d_callables, sample_factor_name) {
  d %>% 
    filter(name == sample_factor_name) %>% 
    rename(sample_factor = name) %>% 
    mutate(
      base = str_extract(type, '^[A-Z]{1}')
    ) %>% 
    
    # join
    left_join(
      d_callables, by = c('sample_factor', 'base')
    ) %>% 
    
    # get rate
    mutate(
      rate = count / callables
    )
}

# MA - average of 940 gens across all lines
rate_summarised$ma = make_final_count_table(count_table, ma_callables_long, 'ma') %>% 
  mutate(callables = callables * 940) %>%  # mean gens in these lines
  mutate(rate = count / callables)

# HS - 150 gens since SL1 + SL2 were crossed to 2935
rate_summarised$hs = make_final_count_table(count_table, hs_callables_long, 'adaptation') %>% 
  mutate(callables = callables * 150) %>% 
  mutate(rate = count / callables)

# salt MA - 236 gens on average across the 5 lines
rate_summarised$salt_0 = make_final_count_table(
  count_table, 
  base_callables_summarised %>% 
    mutate(sample_factor = paste('salt', sample_type, sep = '_')) %>% 
    filter(sample_factor == 'salt_0') %>% 
    select(-sample_type) %>% 
    mutate(callables = callables * 236), 
  'salt_0')
rate_summarised$salt_5 = make_final_count_table(
  count_table, 
  base_callables_summarised %>% 
    mutate(sample_factor = paste('salt', sample_type, sep = '_')) %>% 
    filter(sample_factor == 'salt_5') %>% 
    select(-sample_type) %>% 
    mutate(callables = callables * 236),
  'salt_5')

rate_summarised

rate_summarised_combined = map_df(rate_summarised, bind_rows)

```



## Plots

### Counts

```{r}
obs_exp %>% 
  mutate(
    xcol = str_replace(type, '>', ' \u2192 '),
    name = case_when(
      # name == 'adaptation_nonsilent' ~ 'HS lines (non-silent sites)',
      # name == 'adaptation_silent' ~ 'HS lines (silent sites)',
      name == 'adaptation' ~ 'HS lines',
      name == 'ma' ~ 'MA',
      name == 'salt_0' ~ 'Salt MA (control)',
      name == 'salt_5' ~ 'Salt MA (salt)'
    )) %>% 
  mutate(
    name = fct_relevel(name, 
      c('MA', 'HS lines', 
        'Salt MA (control)', 'Salt MA (salt)'))
  ) %>% 
ggplot(
  aes(x = xcol, y = obs_exp, fill = name)) +
  geom_bar(
    stat = 'identity', 
    position = position_dodge(),
    color = 'black') +
  geom_hline(aes(yintercept = 1), linetype = 'dashed') +
  coord_cartesian(
    y = c(0, 2.5)
  ) +
  theme_classic() +
  labs(
    x = 'Base change type',
    y = 'Observed over expected'
  ) +
  theme(
    axis.text = element_text(size = 12, color = 'black'),
    axis.title = element_text(size = 12, color = 'black'),
    legend.title = element_blank(),
    # legend.position = c(0.2, 0.88),
    legend.position = 'top',
    legend.background = element_blank(),
    legend.text = element_text(size = 10, color = 'black')
  ) +
  guides(
    fill = guide_legend(ncol = 2, nrow = 3)
  ) +
  scale_fill_manual(values = wes_palette(7, name = 'Darjeeling1', type = 'continuous')[c(3,5,6,7)])


```

Separating the MA/HS and individual groups for salt MA:

```{r}
palette = wes_palette(12, name = 'Darjeeling1', type = 'continuous')

obs_exp_salt %>% 
  mutate(
    xcol = str_replace(type, '>', ' \u2192 ')) %>% 
ggplot(
  aes(x = xcol, y = obs_exp, fill = name)) +
  geom_bar(
    stat = 'identity', 
    position = position_dodge(),
    color = 'black') +
  geom_hline(aes(yintercept = 1), linetype = 'dashed') +
  coord_cartesian(
    y = c(0, 2.5)
  ) +
  theme_classic() +
  labs(
    x = 'Base change type',
    y = 'Observed over expected'
  ) +
  theme(
    axis.text = element_text(size = 12, color = 'black'),
    axis.title = element_text(size = 12, color = 'black'),
    legend.title = element_blank(),
    # legend.position = c(0.2, 0.88),
    legend.position = 'top',
    legend.background = element_blank(),
    legend.text = element_text(size = 10, color = 'black')
  ) +
  guides(
    fill = guide_legend(ncol = 3, nrow = 2)
  ) +
  scale_fill_manual(
    values = c(palette[4], palette[5], # CC
               palette[7], palette[9], # DL 
               palette[11], palette[12])) # SL

```

### Rates

Rate plot:

```{r}
palette = wes_palette(12, name = 'Darjeeling1', type = 'continuous')

salt_group_rate %>% 
  mutate(
    type = str_replace(type, '>', ' \u2192 ')) %>% 
ggplot(
  aes(x = type, y = rate, fill = sample_factor)) +
  geom_bar(
    stat = 'identity', 
    position = position_dodge(),
    color = 'black') +
  theme_classic() +
  labs(
    x = 'Base change type',
    y = 'Mutation rate'
  ) +
  theme(
    axis.text = element_text(size = 12, color = 'black'),
    axis.title = element_text(size = 12, color = 'black'),
    legend.title = element_blank(),
    # legend.position = c(0.2, 0.88),
    legend.position = 'top',
    legend.background = element_blank(),
    legend.text = element_text(size = 10, color = 'black')
  ) +
  guides(
    fill = guide_legend(ncol = 3, nrow = 2)
  ) +
  scale_fill_manual(
    values = c(palette[4], palette[5], # CC
               palette[7], palette[9], # DL 
               palette[11], palette[12])) # SL
```

Summarised over 0 and 5:

```{r}
palette = wes_palette(12, name = 'Darjeeling1', type = 'continuous')

salt_rate_summarised %>% 
  mutate(
    type = str_replace(type, '>', ' \u2192 ')) %>% 
ggplot(
  aes(x = type, y = rate, fill = sample_type)) +
  geom_bar(
    stat = 'identity', 
    position = position_dodge(),
    color = 'black') +
  theme_classic() +
  labs(
    x = 'Base change type',
    y = 'Rate'
  ) +
  theme(
    axis.text = element_text(size = 12, color = 'black'),
    axis.title = element_text(size = 12, color = 'black'),
    legend.title = element_blank(),
    # legend.position = c(0.2, 0.88),
    legend.position = 'top',
    legend.background = element_blank(),
    legend.text = element_text(size = 10, color = 'black')
  ) +
  guides(
    fill = guide_legend(ncol = 3, nrow = 2)
  ) +
  scale_fill_manual(
    values = c(palette[11], palette[12])) 

```

Overall, including MA/HS/saltMA:

Get Poisson CIs:

```{r}
get_pois_ci = function(d) {
  count = d$count[1]
  callables = d$callables[1]
  lambda = d$lambda[1]
  
  out = broom::tidy(
    poisson.test(
      x = count, callables, 
      r = lambda, conf.level = 0.95)
  )
  return(out)
}

rate_summarised_cis = list()
for (sample_factor in c('ma', 'hs', 'salt_5')) {
  rate_summarised_cis[[sample_factor]] = rate_summarised[[sample_factor]] %>% 
    group_by(base) %>% 
    mutate(lambda = mean(count) / callables) %>% 
    ungroup() %>% 
    
    # get CIs
    split(.$type) %>%
    map_dfr(
      ~ get_pois_ci(.),
      .id = 'type'
    )
}

rate_summarised_cis = bind_rows(
  rate_summarised_cis, .id = 'sample_factor'
) %>% 
  mutate(
    sample_factor = case_when(
      sample_factor == 'hs' ~ 'HS lines',
      sample_factor == 'ma' ~ 'MA',
      sample_factor == 'salt_5' ~ 'MA in salt'
    )) %>% 
  mutate(
    sample_factor = fct_relevel(sample_factor, 
      c('MA', 'HS lines', 'MA in salt')))


```

Plot with CIs:

```{r}
spectrum_rate_plot_abs = rate_summarised_combined %>% 
  filter(sample_factor != 'salt_0') %>% 
  mutate(
    sample_factor = case_when(
      sample_factor == 'adaptation' ~ 'HS lines',
      sample_factor == 'ma' ~ 'MA',
      sample_factor == 'salt_5' ~ 'MA in salt'
    )) %>% 
  mutate(
    sample_factor = fct_relevel(sample_factor, 
      c('MA', 'HS lines', 'MA in salt'))
  ) %>% 
  mutate(
    # note - eps/pdf can't handle the arrow character, even if it's written directly
    # will need to export svg
    type = str_replace(type, '>', ' → ')) %>% 
  
ggplot(aes(fill = sample_factor)) +
  geom_bar(
    mapping = aes(x = type, y = log10(rate) + 12, fill = sample_factor),
    stat = 'identity', 
    position = position_dodge(),
    color = 'black') +
  geom_errorbar(
    data = rate_summarised_cis,
    mapping = aes(
      x = type,
      ymin = log10(conf.low) + 12, 
      ymax = log10(conf.high) + 12),
    position = position_dodge(width = 0.9),
    width = 0.5
  ) +
  theme_classic() +
  labs(
    x = 'Base change type',
    y = 'Rate'
  ) +
  theme(
    axis.text = element_text(size = 12, color = 'black'),
    axis.title = element_text(size = 12, color = 'black'),
    legend.title = element_blank(),
    # legend.position = c(0.2, 0.88),
    legend.position = 'top',
    legend.background = element_blank(),
    legend.text = element_text(size = 10, color = 'black')
  ) +
  guides(
    fill = guide_legend(ncol = 3, nrow = 2)
  ) +
  scale_fill_manual(values = wes_palette(4, name = 'Darjeeling1', type = 'continuous')[2:4]) +
  scale_y_continuous(
    breaks = seq(0, 4),
    labels = c(
      expression(10^-12), expression(10^-11), expression(10^-10),
      expression(10^-9), expression(10^-8))
  )

spectrum_rate_plot_abs

# ggsave(here('plots/spectrum_rate_plot_abs.svg'), plot = spectrum_rate_plot_abs, width = 8, height = 4)

```

Create normalized plot (each bar divided by the average rate in that group)

```{r}
# need to transform both the data *and* the CIs
# it'll be easiest to combine the CI data frame with the original data for this

label_parse = function(breaks) {
   parse(text = breaks)
}

rate_summarised_combined_clean = rate_summarised_combined %>% 
  filter(sample_factor != 'salt_0') %>% 
  mutate(
    sample_factor = case_when(
      sample_factor == 'adaptation' ~ 'HS',
      sample_factor == 'ma' ~ 'MA',
      sample_factor == 'salt_5' ~ 'Salt MA'
    )) %>% 
  
  # add CIs
  left_join(
    rate_summarised_cis %>% 
      mutate(sample_factor = case_when(
        sample_factor == 'HS lines' ~ 'HS',
        sample_factor == 'MA in salt' ~ 'Salt MA',
        sample_factor == 'MA' ~ 'MA'
      )),
    by = c('sample_factor', 'type')
  ) %>% 
  
  # reorder
  mutate(
    sample_factor = fct_relevel(sample_factor, 
      c('MA', 'HS', 'Salt MA'))
  ) %>% 
  
  # select necessary columns and do transformation
  mutate(
    type = str_replace(type, '>', '  %->% ')) %>% 
  select(
    sample_factor, type, base, rate, conf.low, conf.high
  ) %>% 
  group_by(sample_factor, base) %>% 
  mutate(
    across(
      where(is.numeric), .fns = ~ . / mean(.),
      .names = '{.col}_relative'
    )
  )
  
spectrum_rate_plot_relative = ggplot(rate_summarised_combined_clean, aes(fill = sample_factor)) +
  geom_bar(
    mapping = aes(x = type, y = rate_relative, fill = sample_factor),
    stat = 'identity', 
    position = position_dodge(),
    color = 'black') +
  geom_errorbar(
    mapping = aes(
      x = type,
      ymin = conf.low_relative,
      ymax = conf.high_relative),
    position = position_dodge(width = 0.9),
    width = 0.5
  ) +
  theme_classic() +
  labs(
    x = 'Mutation',
    y = 'Observed / Expected'
  ) +
  theme(
    axis.text = element_text(size = 12, color = 'black', family = 'Helvetica'),
    axis.title = element_text(size = 12, color = 'black'),
    legend.title = element_blank(),
    # legend.position = c(0.2, 0.88),
    legend.position = 'top',
    legend.background = element_blank(),
    legend.text = element_text(size = 10, color = 'black')
  ) +
  guides(
    fill = guide_legend(ncol = 3, nrow = 2)
  ) +
  scale_fill_manual(values = wes_palette(5, name = 'Darjeeling1', type = 'continuous')[c(1, 3, 5)]) +
  coord_cartesian(
    y = c(0, 2)
  ) +
  geom_hline(yintercept = 1, linetype = 'dashed') +
  guides(fill = guide_legend(nrow = 1, ncol = 3)) +
  scale_x_discrete(labels = label_parse)

spectrum_rate_plot_relative

ggsave(here('plots/spectrum_rate_plot_rel.eps'), plot = spectrum_rate_plot_relative, width = 8, height = 4)
```



```{r}
palette = wes_palette(12, name = 'Darjeeling1', type = 'continuous')

rate_summarised_combined %>% 
  mutate(
    sample_factor = case_when(
      # sample_factor == 'adaptation_nonsilent' ~ 'HS lines (non-silent sites)',
      # sample_factor == 'adaptation_silent' ~ 'HS lines (silent sites)',
      sample_factor == 'adaptation' ~ 'HS lines',
      sample_factor == 'ma' ~ 'MA',
      sample_factor == 'salt_0' ~ 'Salt MA (control)',
      sample_factor == 'salt_5' ~ 'Salt MA (salt)'
    )) %>% 
  mutate(
    sample_factor = fct_relevel(sample_factor, 
      c('MA', 'HS lines', 'Salt MA (control)', 'Salt MA (salt)'))
  ) %>% 
  mutate(
    type = str_replace(type, '>', ' \u2192 ')) %>% 
  mutate(
    rate = ifelse(sample_factor != 'MA', rate / 236, rate)
  ) %>% 
  
ggplot(
  aes(x = type, y = rate, fill = sample_factor)) +
  geom_bar(
    stat = 'identity', 
    position = position_dodge(),
    color = 'black') +
  theme_classic() +
  labs(
    x = 'Base change type',
    y = 'Rate'
  ) +
  theme(
    axis.text = element_text(size = 12, color = 'black'),
    axis.title = element_text(size = 12, color = 'black'),
    legend.title = element_blank(),
    # legend.position = c(0.2, 0.88),
    legend.position = 'top',
    legend.background = element_blank(),
    legend.text = element_text(size = 10, color = 'black')
  ) +
  guides(
    fill = guide_legend(ncol = 3, nrow = 2)
  ) +
  scale_fill_manual(values = wes_palette(7, name = 'Darjeeling1', type = 'continuous')[3:7]) 
  scale_y_log10(
    breaks = scales::trans_breaks("log10", function(x) 10^x), 
    labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  coord_cartesian(y = c(10^-12, 10^-8))
  # scale_y_continuous(breaks = scales::log10_trans())
  # scale_y_continuous(trans = 'log10') +

# TODO: Poisson-based CIs for these - num mutations + number of base-based sitegens
# if we redid this, what's the distribution of obs mutations if we expect mean k?


```

Observed over expected - 

```{r}
rate_obs_exp = rate_summarised_combined %>% 
  rename(obs_rate = rate) %>% 
  
  # get exp count
  group_by(sample_factor, base) %>% 
  mutate(exp_muts = round(sum(count) / 3)) %>% 
  mutate(exp_rate = exp_muts / callables) %>% 
  
  # get obs/exp and reorder
  mutate(obs_exp = obs_rate / exp_rate) %>% 
  select(sample_factor, type, obs_exp, everything())

rate_obs_exp
```

Obs/exp plot:

```{r}
rate_obs_exp %>% 
  mutate(
    sample_factor = factor(case_when(
      sample_factor == 'adaptation_nonsilent' ~ 'HS lines (non-silent sites)',
      sample_factor == 'adaptation_silent' ~ 'HS lines (silent sites)',
      sample_factor == 'ma' ~ 'MA',
      sample_factor == 'salt_0' ~ 'Salt MA (control)',
      sample_factor == 'salt_5' ~ 'Salt MA (salt)'
    ))) %>% 
  mutate(
    sample_factor = fct_relevel(sample_factor, 
      c('MA', 'HS lines (non-silent sites)', 'HS lines (silent sites)', 
        'Salt MA (control)', 'Salt MA (salt)'))
  ) %>% 
  mutate(
    type = str_replace(type, '>', ' \u2192 ')) %>% 
  
  ggplot(
    aes(x = type, y = obs_exp, fill = sample_factor)
  ) +
  geom_bar(
    stat = 'identity',
    position = 'dodge',
    colour = 'black'
  ) +
  geom_hline(aes(yintercept = 1), linetype = 'dashed') +
  coord_cartesian(
    y = c(0, 2.5)
  ) +
  theme_classic() +
  labs(
    x = 'Base change type',
    y = 'Observed over expected'
  ) +
  theme(
    axis.text = element_text(size = 12, color = 'black'),
    axis.title = element_text(size = 12, color = 'black'),
    legend.title = element_blank(),
    # legend.position = c(0.2, 0.88),
    legend.position = 'top',
    legend.background = element_blank(),
    legend.text = element_text(size = 10, color = 'black')
  ) +
  guides(
    fill = guide_legend(ncol = 2, nrow = 3)
  ) +
  scale_fill_manual(values = wes_palette(7, name = 'Darjeeling1', type = 'continuous')[3:7])

# looks the exact same 
```



# Equilibrium GC content

$$ \frac{n_{AT -> GC}}{n_{GC -> AT} + n_{AT -> GC}} $$

where $n_{AT -> GC}$ is a 'W->S' mutation, and the reverse is a 'S->W' mutation 
(Mugal, Weber, Ellegren 2015)

```{r}
# GC 68891655 
# AT 38724292
# total 107615947

# GC 37413853 
# AT 20922472
# total 58336325

get_eq_gc = function(d, silent = FALSE, values = FALSE) {
  if (!silent) {
    num_GC = 68891655
    num_AT = 38724292
  } else if (silent) {
    num_GC = 37413853
    num_AT = 20922472
  }
  ws_mutations = c('A>G', 'A>C', 'T>G', 'T>C')
  sw_mutations = c('G>A', 'G>T', 'C>A', 'C>T')
  
  n_ws = d %>% 
    filter(mutation %in% ws_mutations) %>% 
    nrow()
  
  n_sw = d %>% 
    filter(mutation %in% sw_mutations) %>% 
    nrow()
  
  n_ws_norm = n_ws / num_AT
  n_sw_norm = n_sw / num_GC
  
  if (values) {
    return(c(
      'n_ws' = n_ws,
      'n_sw' = n_sw,
      'num_AT' = num_AT,
      'num_GC' = num_GC))
  } else if (!values) {
    eq_gc = n_ws_norm / (n_ws_norm + n_sw_norm)
    message(paste(n_ws, n_sw))
    message(paste(n_ws / num_AT, n_sw / num_GC))
    return(eq_gc)
  }
}
```

Calculate values:

```{r}
get_eq_gc(ma)

get_eq_gc(adaptation_silent_nonsilent %>% filter(site_type == 'silent'), silent = TRUE)

map(
  salt %>% split(.$sample_type),
  ~ get_eq_gc(.)
)
```



## WS and SW mutation counts

```{r}
# MA orig - mutations to ATs vs mutations to GCs
ma_orig_values = get_eq_gc(ma, values = TRUE)
message('ma')
as.numeric(ma_orig_values['n_sw'] / ma_orig_values['n_ws']) # 4.65

# adaptation - silent and nonsilent
adaptation_silent_values = get_eq_gc(
  adaptation_silent_nonsilent %>% 
    filter(site_type == 'silent'), 
  values = TRUE)
adaptation_nonsilent_values = get_eq_gc(
  adaptation_silent_nonsilent %>% 
    filter(site_type == 'nonsilent'), 
  values = TRUE)
adaptation_all_values = get_eq_gc(
  adaptation_silent_nonsilent,
  values = TRUE
)

message('adaptation silent')
as.numeric(adaptation_silent_values['n_sw'] / adaptation_silent_values['n_ws']) # silent - 3.65
message('adaptation nonsilent')
as.numeric(adaptation_nonsilent_values['n_sw'] / adaptation_silent_values['n_ws']) # nonsilent - 3.75
message('adaptation all')
as.numeric(adaptation_all_values['n_sw'] / adaptation_all_values['n_ws']) # all - 3.06

# salt ma - across both types
map(
  salt %>% split(.$sample_type),
  function(d) {
    vals = get_eq_gc(d, values = TRUE)
    mut_ratio = vals['n_sw'] / vals['n_ws']
    return(as.numeric(mut_ratio))
  }
)


```



# Ts/Tv

```{r}
ts_tv = function(d, values = FALSE) {
  tv = c('A>T', 'T>A', 'G>C', 'C>G', 'A>C', 'C>A', 'G>T', 'T>G')
  ts = c('A>G', 'G>A', 'C>T', 'T>C')
  ts_count = d %>% 
    filter(mutation %in% ts) %>% 
    nrow()
  tv_count = d %>% 
    filter(mutation %in% tv) %>% 
    nrow()
  ts_tv_ratio = ts_count / tv_count
  if (!values) {
    return(ts_tv_ratio)
  } else if (values) {
    return(c('ts' = ts_count, 'tv' = tv_count))
  }
}
```

Running over datasets:

```{r}
message('ma')
ts_tv(ma)

message('adaptation')
ts_tv(adaptation_silent_nonsilent)
ts_tv(adaptation_silent_nonsilent %>% filter(site_type == 'silent'))
ts_tv(adaptation_silent_nonsilent %>% filter(site_type == 'nonsilent'))

message('salt_summarised')
map(
  salt %>% 
    split(.$sample_type),
  ~ ts_tv(.)
)

message('salt_ma')
map(
  salt %>% 
    split(.$sample_factor),
  ~ ts_tv(.)
)
```

Binomial tests - classifying Ts as 'success', meaning prob(success) is 1/3 -

```{r}
# original MA
ma_vals = ts_tv(ma, values = TRUE)
binom.test(ma_vals['ts'], sum(ma_vals), alternative = 'two.sided', conf.level = 0.95) 

d_list = list(
  'ma' = ma, 
  'adaptation' = adaptation_silent_nonsilent,
  'adaptation_silent' = adaptation_silent_nonsilent %>% filter(site_type == 'silent'),
  'adaptation_nonsilent' = adaptation_silent_nonsilent %>% filter(site_type == 'nonsilent'))

# salt_list = salt %>% split(.$sample_factor)
salt_list = salt %>% split(.$sample_type)

d_list = append(d_list, salt_list)

ts_tv_all = map_dfr(
  d_list,
  function(d) {
    vals = ts_tv(d, values = TRUE)
    out = binom.test(
      vals['ts'], sum(vals), p = 1/3,
      alternative = 'two.sided',
      conf.level = 0.95
    ) %>% 
      broom::tidy()
  },
  .id = 'sample_factor'
)

ts_tv_all

```

Comparing Ts/Tv across groups:

```{r}
message('MA vs HS')
fisher.test(
  rbind(
    ts_tv(ma, values = TRUE),
    ts_tv(adaptation, values = TRUE)
  )
)

message('HS vs Salt MA')
fisher.test(
  rbind(
    ts_tv(adaptation, values = TRUE),
    ts_tv(salt %>% filter(sample_type == '5'), values = TRUE)
  )
)

message('MA vs Salt MA')
fisher.test(
  rbind(
    ts_tv(ma, values = TRUE),
    ts_tv(salt %>% filter(sample_type == '5'), values = TRUE)
  )
)
```

Proportion of C>T over other mut types:

```{r}
count_table %>% 
  group_by(name) %>% 
  mutate(total = sum(count)) %>% 
  filter(type == 'C>T') %>% 
  mutate(CT_prop = count / total)
```

