---
title: "KaKs"
author: Ahmed Hasan
output: pdf_notebook
---

# Package import

```{r}
library(tidyverse)
library(fs)
library(here)
library(broom)
```

# Data import

```{r}
all_genome = read_tsv(here('data/rate/ka_ks/all_genome.tsv'), col_types = cols()) %>% 
  mutate(sample_type = case_when(
    str_ends(sample, '0') ~ 0,
    str_ends(sample, '5') ~ 5
  )) %>% 
  mutate(sample_group = str_extract(sample, '^[A-Z]{2}'))

all_genome_callables = read_tsv(here('data/rate/ka_ks/degen_callables_lookup.tsv'), col_types = cols()) %>% 
  mutate(sample_type = case_when(
    str_ends(sample, '0') ~ 0,
    str_ends(sample, '5') ~ 5
  )) %>% 
  mutate(sample_group = str_extract(sample, '^[A-Z]{2}'))

all_perrineau = read_tsv(
  here('data/rate/ka_ks/syn_nonsyn_perrineau_counts.tsv'), col_types = cols()) %>% 
  mutate(total_muts = rowSums(
    across(contains('muts')))
  ) %>% 
  filter(total_muts > 0) %>%
  select(-total_muts) 
  
all_perrineau_callables = read_tsv(
  here('data/rate/ka_ks/perrineau_genes_callables.tsv'), col_types = cols())

all_expressed = read_tsv(
  here('data/rate/ka_ks/syn_nonsyn_expressed_counts.tsv'), col_types = cols()) %>% 
  mutate(total_muts = rowSums(
    across(contains('muts')))
  ) %>% 
  filter(total_muts > 0) %>%  # 41 mutated genes
  select(-total_muts) 
  
all_expressed_callables = read_tsv(
  here('data/rate/ka_ks/expressed_genes_callables.tsv'), col_types = cols())
```

# Ka/Ks calc

## Genomewide

Need to calculate:

$$

\frac{K_a}{K_s} = \frac{\text{nonsyn muts} / \text{nonsyn callables}}{\text{syn muts} / \text{syn callables}}

$$

```{r}
# get callables per sample type
genome_sample_callables = all_genome_callables %>% 
  # filter(str_detect(scaffold, 'scaffold', negate = TRUE)) %>% 
  group_by(sample_group, sample_type) %>% 
  summarise(across(contains('syn'), sum))

# get muts per sample type
genome_sample_counts = all_genome %>% 
  select(-nonsyn, -syn) %>% # these values only consider chrs with muts
  group_by(sample_group, sample_type) %>% 
  summarise(across(.cols = contains('syn'), .fns = sum))
```

and now to calculate Ka/Ks:

```{r}
genome_sample_counts %>% 
  
# join callables and counts
  full_join(
    genome_sample_callables, by = c('sample_group', 'sample_type')
  ) %>% 
  replace_na(list(nonsyn_muts = 0, syn_muts = 0)) %>% # SL5 has none

# compute Ka, Ks, and Ka/Ks values for 0 and 5
  mutate(
    ka = nonsyn_muts / nonsyn,
    ks = syn_muts / syn
  ) %>% 
  mutate(
    ka_ks = ka / ks
  )
```

## Bootstrap confidence intervals

Need to bootstrap original mutation dataset for this -

### Load in and prepare data

```{r}
muts_described = read_tsv(
  here('data/mutations/mut_describer/muts_described.final.tsv'), 
  col_types = cols()
) %>% 
  
  # only keep cols of interest and remove sites not classed as syn/nonsyn
  select(
    chromosome, position, mutant_sample, nonsyn_v_syn
  ) %>% 
  filter(!is.na(nonsyn_v_syn)) %>% 
  
  # add sample group and type as cols
  mutate(
    sample_group = str_extract(mutant_sample, '^[A-Z]{2}'),
    sample_type = str_extract(mutant_sample, '[05]$')
  )
```

Resampling muts with replacement:

```{r}
ka_ks_calc = function(run_id) {
  # resample (w/ replacement) mutations in each sample group + type - 6 total groups
  d_resampled = muts_described %>% 
    group_by(sample_group, sample_type) %>% 
    slice_sample(prop = 1, replace = TRUE)
  
  # compute ka/ks value for each
  d_summarised = d_resampled %>% 
    
    # get counts of nonsyn and syn muts
    group_by(sample_group, sample_type) %>% 
    count(nonsyn_v_syn) %>% 
    pivot_wider(
      names_from = nonsyn_v_syn,
      values_from = n,
      values_fill = 0 # SL syn
    ) %>% 
    # some cleanup
    rename(nonsyn_muts = nonsynonymous, syn_muts = synonymous) %>% 
    mutate(sample_type = as.numeric(sample_type)) %>% 
    
    # get ka/ks
    full_join(
      genome_sample_callables, by = c('sample_group', 'sample_type')
    ) %>% 
      replace_na(
        list(nonsyn_muts = 0, syn_muts = 0)
    ) %>% 
    mutate(
      ka = nonsyn_muts / nonsyn,
      ks = syn_muts / syn
    ) %>%
    mutate(
      ka_ks = ka / ks) %>% 
    
    # remove extra columns and prep for 'export'
    select(sample_group, sample_type, ka_ks) %>% 
    ungroup() %>% 
    mutate(run_id = run_id)

    return(d_summarised)
}

ka_ks_boot_vals = map_dfr(1:1000, ~ ka_ks_calc(.)) # runs for ~30 seconds

```

Calculate confidence intervals from bootstrap vals:

```{r}
boot_vals_list = ka_ks_boot_vals %>% 
  filter(sample_group != 'SL') %>% # all inf/NA due to lack of data
  mutate(sample_factor = paste(sample_group, sample_type, sep = '_')) %>% 
  split(.$sample_factor)

get_cis = function(d) {
  d_id = d %>% 
    select(sample_factor) %>% 
    unlist() %>% 
    unique() 
  
  boot_vals = d %>% 
    select(ka_ks) %>% 
    filter(!is.na(ka_ks), is.finite(ka_ks)) %>% 
    unlist() %>% 
    t.test(., alternative = 'two.sided', mu = 0, conf.level = 0.95) %>% 
    tidy()
  
  boot_vals$sample_factor = d_id
  boot_vals = select(boot_vals, sample_factor, conf.low, conf.high, everything())
  
  return(boot_vals)
}

map_dfr(
  boot_vals_list,
  ~ get_cis(.)
)

```


## Expressed genes

How many genes had mutations in multiple samples? 

```{r}
all_expressed %>% 
  group_by(gene_name) %>% 
  count() %>% 
  arrange(-n) # just NP_042564.1
```

Creating a gene name column:

```{r}
all_genes = unique(all_expressed$gene_name)
all_gene_names = all_expressed_callables$gene_name
g_temp = character(length(all_gene_names))

get_gene = function(string, pattern) {
  gene_vec = str_extract(string, pattern)
  gene_name = gene_vec[!is.na(gene_vec)]
  return(gene_name)
}

# pain
# no idea why mutate() refused to work for this operation
for (i in seq_along(g_temp)) {
  gene = get_gene(all_gene_names[i], all_genes)
  g_temp[i] = gene
}

all_expressed_callables$actual_gene_name = g_temp
```

Joining the callables dataset with the mutation counts and computing Ka/Ks:

```{r}
all_expressed %>% 
  
  # transform to long format and separate by sample type
  pivot_longer(cols = contains('muts'), names_to = 'measure') %>% 
  mutate(
    sample_type = str_extract(sample, '([05])$'), 
    measure_type = str_extract(measure, '([0-5])$')) %>% 
  mutate( 
    measure = str_replace(measure, '_[05]', '')
  ) %>% 
  filter(sample_type == measure_type) %>% 
  select(-measure_type) %>% 

  # transform back to wide format
  pivot_wider(names_from = measure, values_from = value) %>% 
  
  # combine with similarly sample-type-separated callables file
  full_join(
    all_expressed_callables %>% 
      rename(all_names = gene_name) %>% 
      rename(gene_name = actual_gene_name) %>% 
      select(-all_names),
    by = c('sample', 'gene_name')
  ) %>% 
  
  # some data cleanup post full join
  distinct() %>% # full join induced some duplicates
  arrange(sample) %>% 
  mutate(sample_type = str_extract(sample, '([05])$')) %>%  # for newly added rows 
  mutate(nonsyn_muts = ifelse(is.na(nonsyn_muts), 0, nonsyn_muts),
         syn_muts = ifelse(is.na(syn_muts), 0, syn_muts)) %>% 
  
  # group by sample group and sample type and get sums for mut counts and callable counts
  mutate(sample_group = str_extract(sample, '^[A-Z]{2}')) %>% 
  select(sample, sample_group, sample_type, gene_name, 
         chrom, start, end, everything(), -contains('fold')) %>% 
  group_by(sample_group, sample_type) %>% 
  summarise(across(contains('syn'), sum)) %>% 
  
  # compute Ka, Ks, and Ka/Ks values for 0 and 5 samples across sample groups
  mutate(
    ka = nonsyn_muts / nonsyn,
    ks = syn_muts / syn
  ) %>% 
  mutate(
    ka_ks = ka / ks
  )
```






















