---
title: "Salt lines supplemental"
author: Ahmed Hasan
output: pdf_document
---

# Package import

```{r}
library(tidyverse)
library(here)
library(fs)
library(wesanderson)
```

# Tables

## Table S1 - HS mutations

```{r}
adaptation = read_tsv(
  here('data/rate/mut_describer/adaptation.gene_sets.filtered.tsv'),
  col_types = cols(), na = c('.')
) %>% 
  select(
    chromosome, position, mutation, type, mutant_sample,
    gene = Gene.primaryIdentifier, is_salt_gene = perrineau,
    genic:CDS
  )

write_tsv(adaptation, file = here('data/supplemental/table_s1.tsv'))
```

## Table S2 - Salt MA mutations

```{r}
salt_ma = read_tsv(
  here('data/rate/mut_describer/muts_described.final.gene_sets.tsv'),
  col_types = cols()
) %>% 
  mutate(rho = as.double(rho)) %>% 
  bind_rows(
    read_tsv(
      here('data/rate/mut_describer/indels_described.gene_sets.tsv'),
      col_types = cols()
    ) %>% 
      mutate(mut_codon = as.character(mut_codon)) # needed for join
  ) %>% 
  mutate(
    sample_group = str_extract(mutant_sample, '^[A-Z]{2}'),
    sample_type = str_extract(mutant_sample, '[05]$'),
    sample_factor = paste(sample_group, sample_type, sep = '_')
  ) %>% 
  filter(sample_type == '5') %>% 
   mutate(
    gene = str_match(feature_names, "^\\[\\'([A-Za-z\\.0-9]+)")[, 2] # lol
  ) %>% 
  select(
    chromosome, position, mutation, type, mutant_sample,
    gene, is_salt_gene = perrineau, genic:CDS
  )

write_tsv(salt_ma, file = here('data/supplemental/table_s2.tsv'))

```

# Figures

## Figure S1 - annotations within Salt MA

Data generated in `annotation.Rmd` - loading in:

```{r}
salt_ma_spatial = read_tsv(
  here('data/spatial/salt_ma_annotations_cis.tsv'),
  col_types = cols()
) %>% 
  select(sample_factor, ant = annotation, prop, contains('ci')) %>% 
  mutate(gene_set = 'genome wide') %>% 
  filter(str_extract(sample_factor, '[05]$') == 5) %>% 
  mutate(
    sample_factor = case_when(
      sample_factor == 'CC_5' ~ 'original MA strains',
      sample_factor == 'DL_5' ~ 'HS line ancestors',
      sample_factor == 'SL_5' ~ 'salt tolerant lines'
    )
  )

salt_ma_spatial_perrineau = read_tsv(
  here('data/supplemental/salt_ma_perrineau.tsv'),
  col_types = cols()
) %>% 
  mutate(gene_set = 'candidate salt genes') %>% 
  filter(!is.na(prop)) %>% 
  filter(str_extract(sample_factor, '[05]$') == 5) %>% 
  mutate(
    sample_factor = case_when(
      sample_factor == 'CC_5' ~ 'original MA strains',
      sample_factor == 'DL_5' ~ 'HS line ancestors',
      sample_factor == 'SL_5' ~ 'salt tolerant lines'
    )
  )
```

Plot (code from `annotation.Rmd`)

```{r}
palette = wes_palette('Darjeeling1', 5)

fig_s1 = salt_ma_spatial %>% 
  bind_rows(salt_ma_spatial_perrineau) %>% 
  # set up data
  mutate(sample_factor = ifelse(sample_factor == 'Salt MA (salt)', 'Salt MA', sample_factor),
         gene_set = ifelse(gene_set == 'salt genes', 'candidate salt genes', gene_set)) %>% 
  mutate(
    ant = case_when(
      ant == 'utr5' ~ "5' UTR",
      ant == 'utr3' ~ "3' UTR",
      TRUE ~ as.character(ant)
    )
  ) %>% 
  mutate(
    sample_factor = fct_relevel(sample_factor, 
      c('original MA strains', 'HS line ancestors', 'salt tolerant lines')),
    gene_set = fct_relevel(
      gene_set, c('genome wide', 'candidate salt genes')),
    ant = fct_relevel(
      ant, c('intergenic', "5' UTR", 'CDS', 'intronic', "3' UTR")
    )
  ) %>% 

  # plot
  ggplot(
    aes(x = ant, y = prop, fill = sample_factor)) +
  geom_bar(
    stat = 'identity', 
    position = position_dodge(width = 0.9),
    colour = 'black',
    size = 0.4) +
  geom_errorbar(
    aes(ymin = lower_ci, ymax = upper_ci),
    position = position_dodge(width = 0.9),
    size = 0.4, width = 0.2
  ) +
  facet_grid(
    ~ gene_set, scales = 'free') +
  coord_cartesian(
    y = c(0, 1)
  ) +
  labs(
    x = '',
    y = 'Proportion of all mutations'
  ) +
  scale_fill_manual(
    values = c(
      'original MA strains' = palette[1], # '#87C5FC', orig color
      'HS line ancestors' = palette[3], # '#FF5B5F', orig color
      # 'Salt MA (control)' = palette[6], 
      'salt tolerant lines' = palette[5])
  ) +
  # macro elements
  guides(fill = guide_legend(
    override.aes = list(color = 'white'))
  ) +
  theme(
    panel.border = element_rect(colour = 'black', size = 0.5, fill = NA),
    panel.background = element_blank(),
    strip.background = element_blank(),
    legend.position = c(0.175, 0.8),
    legend.text = element_text(size = 8, colour = 'black'),
    legend.title = element_blank(),
    legend.key = element_rect(color = 'white'),
    legend.key.size = unit(0.5, 'cm')) +
  # text elements
  theme(
    strip.text = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.title = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.text = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

fig_s1

ggsave(here('plots/fig_s1.eps'), plot = fig_s1, width = 6, height = 4)
```

## Figure S2 - Ka/Ks across salt MA groups

File generated in `analysis/rate/ka_ks.Rmd`:

```{r}
ka_ks = read_tsv(
  here('data/rate/ka_ks/ka_ks_genome_final_CIs.tsv'),
  col_types = cols()
)
```

Plot (code mostly from `analysis/rate/ka_ks.Rmd`)

```{r}
palette = wes_palette(5, name = 'Darjeeling1')

# relevel
fig_s2 = ka_ks %>% 
  filter(sample_type == '5') %>% 
  mutate(
    sample_group = case_when(
      sample_group == 'CC' ~ 'original MA strains',
      sample_group == 'DL' ~ 'HS line ancestors',
      sample_group == 'SL' ~ 'salt tolerant lines')) %>% 
  mutate(
    sample_group = fct_relevel(
      sample_group, 
      c('original MA strains', 'HS line ancestors', 'salt tolerant lines'))) %>% 
  
  # plot
  ggplot(
    aes(x = sample_group, y = ka_ks, fill = sample_group)
  ) +
  geom_bar(
    stat = 'identity',
    color = 'black',
    size = 0.4
  ) +
  geom_errorbar(
    aes(ymin = lower_ci, ymax = upper_ci),
    size = 0.4, width = 0.2
  ) +
  geom_hline(
    yintercept = 1,
    linetype = 'dashed',
    size = 0.5,
    alpha = 1,
    color = 'black'
  ) +
  labs(
    x = '',
    y = expression(paste(K[a] / K[s], ' ratio'))
  ) +
  theme(
    panel.border = element_rect(colour = 'black', size = 0.5, fill = NA),
    panel.background = element_blank(),
    strip.background = element_blank(),
    legend.position = c(0.85, 0.75),
    legend.text = element_text(size = 8, colour = 'black'),
    legend.title = element_blank(),
    legend.key = element_rect(color = 'white'),
    legend.key.size = unit(0.5, 'cm')) +
  # text elements
  theme(
    strip.text = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.title = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.text = element_text(family = 'Helvetica', colour = 'black', size = 10),
    axis.text.x = element_text(angle = 45, hjust = 1)
  ) +
  scale_fill_manual(
    values = palette[c(1, 3, 5)]
  ) +
  scale_y_continuous(
    breaks = seq(0, 6.0, 0.5)
  ) +
  guides(fill = FALSE)

fig_s2

ggsave(here('plots/fig_s2.eps'), plot = fig_s2, width = 3, height = 4)
```

## Figure S3 - intermutation distances

Generated in `analysis/spatial/intermutation_distances.Rmd`:

```{r}
intermut_salt_ma = read_tsv(
  here('data/spatial/saltMA_intermut.tsv'),
  col_types = cols()
)
```

Plot (also mostly taken from `analysis/spatial/intermutation_distances.Rmd`)

```{r}
fig_s3 = intermut_salt_ma %>% 
  filter(sample_group != 'SL') %>% # only two muts
  mutate(
    sample_group = case_when(
      sample_group == 'CC' ~ 'original MA strains',
      sample_group == 'DL' ~ 'HS line ancestors')) %>% 
  mutate(
    sample_group = fct_relevel(
      sample_group, 
      c('original MA strains', 'HS line ancestors'))) %>% 
  
  ggplot(
    aes(x = log10(d), color = sample_group)) +
  stat_ecdf(
    size = 1, 
    geom = 'line') +
  scale_color_manual(
    values = wes_palette(
      5, name = 'Darjeeling1', type = 'continuous')[c(1, 3)]) +
  labs(
    x = expression(
      paste('Intermutation distance (', log[10], ' bp)')),
    y = 'Cumulative density'
  ) +
  # legend styling
  theme(
    legend.text = element_text(family = 'Helvetica', color = 'black', size = 10),
    legend.position = c(0.18, 0.8),
    legend.title = element_blank(),
    legend.box.background = element_blank(),
    legend.key = element_blank()
  ) +
  theme(
    panel.background = element_blank(),
    axis.line = element_line(size = 0.5, color = 'black'),
    axis.ticks = element_line(size = 0.5, color = 'black'),
    axis.title = element_text(family = 'Helvetica', size = 12, color = 'black'),
    axis.text = element_text(family = 'Helvetica', size = 12, color = 'black')
  )

fig_s3

ggsave(here('plots/fig_s3.eps'), plot = fig_s3, width = 6, height = 4)
```

## Figure S4 - mut base spectrum

From `analysis/spectrum_context/mut_base_spectrum.Rmd`

```{r}
salt_spectrum = read_tsv(
  here('data/spectrum_context/saltMA_spectrum.tsv'),
  col_types = cols()
) %>% 
  filter(str_extract(sample_factor, '[05]$') == 5)

get_pois_ci = function(d) {
  count = d$count[1]
  callables = d$callables[1]
  lambda = d$lambda[1]
  
  out = broom::tidy(
    poisson.test(
      x = count, callables, 
      r = lambda, conf.level = 0.95)
  )
  return(out)
}

# get poisson CIs
salt_spectrum_cis = list()
for (sample_factor_current in c('CC_5', 'DL_5', 'SL_5')) {
  salt_spectrum_cis[[sample_factor_current]] = salt_spectrum %>% 
    filter(sample_factor == sample_factor_current) %>% 
    group_by(base) %>% 
    mutate(lambda = mean(count) / callables) %>% 
    ungroup() %>% 
    
    # get CIs
    split(.$type) %>%
    map_dfr(
      ~ get_pois_ci(.),
      .id = 'type'
    )
}

salt_spectrum_cis = bind_rows(
  salt_spectrum_cis, .id = 'sample_factor'
) %>% 
  mutate(
    sample_factor = case_when(
      sample_factor == 'CC_5' ~ 'original MA strains',
      sample_factor == 'DL_5' ~ 'HS line ancestors',
      sample_factor == 'SL_5' ~ 'salt tolerant lines'
    )) %>% 
  mutate(
    sample_factor = fct_relevel(sample_factor, 
      c('original MA strains', 'HS line ancestors', 'salt tolerant lines')))

# used for plotting
label_parse = function(breaks) {
   parse(text = breaks)
}

# prep df for plotting
salt_spectrum_final = salt_spectrum %>% 
  mutate(
    sample_factor = case_when(
      sample_factor == 'CC_5' ~ 'original MA strains',
      sample_factor == 'DL_5' ~ 'HS line ancestors',
      sample_factor == 'SL_5' ~ 'salt tolerant lines'
    )) %>% 
  
  # add CIs
  left_join(salt_spectrum_cis) %>% 
  
  # reorder
  mutate(
    sample_factor = fct_relevel(sample_factor, 
      c('original MA strains', 'HS line ancestors', 'salt tolerant lines'))
  ) %>% 
  
  # select necessary columns and do transformation
  mutate(
    type = str_replace(type, '>', '  %->% ')) %>% 
  select(
    sample_factor, type, base, rate, conf.low, conf.high
  ) %>% 
  group_by(sample_factor, base) %>% 
  mutate(
    across(
      where(is.numeric), .fns = ~ . / mean(.),
      .names = '{.col}_relative'
    )
  )

salt_spectrum_final
```

Plot:

```{r}
fig_s4 = ggplot(salt_spectrum_final, aes(fill = sample_factor)) +
  geom_bar(
    mapping = aes(x = type, y = rate_relative, fill = sample_factor),
    stat = 'identity', 
    position = position_dodge(),
    color = 'black') +
  geom_errorbar(
    mapping = aes(
      x = type,
      ymin = conf.low_relative,
      ymax = conf.high_relative),
    position = position_dodge(width = 0.9),
    width = 0.5
  ) +
  theme_classic() +
  labs(
    x = 'Mutation',
    y = 'Observed / Expected'
  ) +
  theme(
    axis.text = element_text(size = 12, color = 'black', family = 'Helvetica'),
    axis.title = element_text(size = 12, color = 'black'),
    legend.title = element_blank(),
    # legend.position = c(0.2, 0.88),
    legend.position = 'top',
    legend.background = element_blank(),
    legend.text = element_text(size = 10, color = 'black')
  ) +
  guides(
    fill = guide_legend(ncol = 3, nrow = 2)
  ) +
  scale_fill_manual(values = wes_palette(5, name = 'Darjeeling1', type = 'continuous')[c(1, 3, 5)]) +
  geom_hline(yintercept = 1, linetype = 'dashed') +
  guides(fill = guide_legend(nrow = 1, ncol = 3)) +
  scale_x_discrete(labels = label_parse)

fig_s4

ggsave(here('plots/fig_s4.eps'), plot = fig_s4, width = 8, height = 4)
```

